// @wordpress/block-library/build-module/image/image-editing/aspect-ratio-dropdown.js
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { check, aspectRatio as aspectRatioIcon } from '@wordpress/icons';
import { DropdownMenu, MenuGroup, MenuItem } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import { POPOVER_PROPS } from './constants';
import { useImageEditingContext } from './context';

function AspectGroup(_ref) {
  var aspectRatios = _ref.aspectRatios,
      isDisabled = _ref.isDisabled,
      label = _ref.label,
      _onClick = _ref.onClick,
      value = _ref.value;
  return createElement(MenuGroup, {
    label: label
  }, aspectRatios.map(function (_ref2) {
    var title = _ref2.title,
        aspect = _ref2.aspect;
    return createElement(MenuItem, {
      key: aspect,
      disabled: isDisabled,
      onClick: function onClick() {
        _onClick(aspect);
      },
      role: "menuitemradio",
      isSelected: aspect === value,
      icon: aspect === value ? check : undefined
    }, title);
  }));
}

export default function AspectRatioDropdown(_ref3) {
  var toggleProps = _ref3.toggleProps;

  var _useImageEditingConte = useImageEditingContext(),
      isInProgress = _useImageEditingConte.isInProgress,
      aspect = _useImageEditingConte.aspect,
      setAspect = _useImageEditingConte.setAspect,
      defaultAspect = _useImageEditingConte.defaultAspect;

  return createElement(DropdownMenu, {
    icon: aspectRatioIcon,
    label: __('Aspect Ratio'),
    popoverProps: POPOVER_PROPS,
    toggleProps: toggleProps,
    className: "wp-block-image__aspect-ratio"
  }, function (_ref4) {
    var onClose = _ref4.onClose;
    return createElement(Fragment, null, createElement(AspectGroup, {
      isDisabled: isInProgress,
      onClick: function onClick(newAspect) {
        setAspect(newAspect);
        onClose();
      },
      value: aspect,
      aspectRatios: [{
        title: __('Original'),
        aspect: defaultAspect
      }, {
        title: __('Square'),
        aspect: 1
      }]
    }), createElement(AspectGroup, {
      label: __('Landscape'),
      isDisabled: isInProgress,
      onClick: function onClick(newAspect) {
        setAspect(newAspect);
        onClose();
      },
      value: aspect,
      aspectRatios: [{
        title: __('16:10'),
        aspect: 16 / 10
      }, {
        title: __('16:9'),
        aspect: 16 / 9
      }, {
        title: __('4:3'),
        aspect: 4 / 3
      }, {
        title: __('3:2'),
        aspect: 3 / 2
      }]
    }), createElement(AspectGroup, {
      label: __('Portrait'),
      isDisabled: isInProgress,
      onClick: function onClick(newAspect) {
        setAspect(newAspect);
        onClose();
      },
      value: aspect,
      aspectRatios: [{
        title: __('10:16'),
        aspect: 10 / 16
      }, {
        title: __('9:16'),
        aspect: 9 / 16
      }, {
        title: __('3:4'),
        aspect: 3 / 4
      }, {
        title: __('2:3'),
        aspect: 2 / 3
      }]
    }));
  });
}
