// @wordpress/block-library/build-module/image/image-editing/cropper.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import Cropper from 'react-easy-crop';
import classnames from 'classnames';
/**
 * WordPress dependencies
 */

import { Spinner } from '@wordpress/components';
/**
 * Internal dependencies
 */

import { MIN_ZOOM, MAX_ZOOM } from './constants';
import { useImageEditingContext } from './context';
export default function ImageCropper(_ref) {
  var url = _ref.url,
      width = _ref.width,
      height = _ref.height,
      clientWidth = _ref.clientWidth,
      naturalHeight = _ref.naturalHeight,
      naturalWidth = _ref.naturalWidth;

  var _useImageEditingConte = useImageEditingContext(),
      isInProgress = _useImageEditingConte.isInProgress,
      editedUrl = _useImageEditingConte.editedUrl,
      position = _useImageEditingConte.position,
      zoom = _useImageEditingConte.zoom,
      aspect = _useImageEditingConte.aspect,
      setPosition = _useImageEditingConte.setPosition,
      setCrop = _useImageEditingConte.setCrop,
      setZoom = _useImageEditingConte.setZoom,
      rotation = _useImageEditingConte.rotation;

  var editedHeight = height || clientWidth * naturalHeight / naturalWidth;

  if (rotation % 180 === 90) {
    editedHeight = clientWidth * naturalWidth / naturalHeight;
  }

  return createElement("div", {
    className: classnames('wp-block-image__crop-area', {
      'is-applying': isInProgress
    }),
    style: {
      width: width || clientWidth,
      height: editedHeight
    }
  }, createElement(Cropper, {
    image: editedUrl || url,
    disabled: isInProgress,
    minZoom: MIN_ZOOM / 100,
    maxZoom: MAX_ZOOM / 100,
    crop: position,
    zoom: zoom / 100,
    aspect: aspect,
    onCropChange: setPosition,
    onCropComplete: function onCropComplete(newCropPercent) {
      setCrop(newCropPercent);
    },
    onZoomChange: function onZoomChange(newZoom) {
      setZoom(newZoom * 100);
    }
  }), isInProgress && createElement(Spinner, null));
}
