// @wordpress/block-library/build-module/image/image-editing/use-save-image.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";

/**
 * WordPress dependencies
 */
import apiFetch from '@wordpress/api-fetch';
import { useDispatch } from '@wordpress/data';
import { useCallback, useMemo, useState } from '@wordpress/element';
import { __, sprintf } from '@wordpress/i18n';
import { store as noticesStore } from '@wordpress/notices';
export default function useSaveImage(_ref) {
  var crop = _ref.crop,
      rotation = _ref.rotation,
      height = _ref.height,
      width = _ref.width,
      aspect = _ref.aspect,
      url = _ref.url,
      id = _ref.id,
      onSaveImage = _ref.onSaveImage,
      onFinishEditing = _ref.onFinishEditing;

  var _useDispatch = useDispatch(noticesStore),
      createErrorNotice = _useDispatch.createErrorNotice;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isInProgress = _useState2[0],
      setIsInProgress = _useState2[1];

  var cancel = useCallback(function () {
    setIsInProgress(false);
    onFinishEditing();
  }, [setIsInProgress, onFinishEditing]);
  var apply = useCallback(function () {
    setIsInProgress(true);
    var attrs = {}; // The crop script may return some very small, sub-pixel values when the image was not cropped.
    // Crop only when the new size has changed by more than 0.1%.

    if (crop.width < 99.9 || crop.height < 99.9) {
      attrs = crop;
    }

    if (rotation > 0) {
      attrs.rotation = rotation;
    }

    attrs.src = url;
    apiFetch({
      path: "/wp/v2/media/".concat(id, "/edit"),
      method: 'POST',
      data: attrs
    }).then(function (response) {
      onSaveImage({
        id: response.id,
        url: response.source_url,
        height: height && width ? width / aspect : undefined
      });
    }).catch(function (error) {
      createErrorNotice(sprintf(
      /* translators: 1. Error message */
      __('Could not edit image. %s'), error.message), {
        id: 'image-editing-error',
        type: 'snackbar'
      });
    }).finally(function () {
      setIsInProgress(false);
      onFinishEditing();
    });
  }, [setIsInProgress, crop, rotation, height, width, aspect, url, onSaveImage, createErrorNotice, setIsInProgress, onFinishEditing]);
  return useMemo(function () {
    return {
      isInProgress: isInProgress,
      apply: apply,
      cancel: cancel
    };
  }, [isInProgress, apply, cancel]);
}
