// @wordpress/block-library/build-module/image/image-editing/zoom-dropdown.js
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { ToolbarButton, RangeControl, Dropdown } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { search } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import { MIN_ZOOM, MAX_ZOOM, POPOVER_PROPS } from './constants';
import { useImageEditingContext } from './context';
export default function ZoomDropdown() {
  var _useImageEditingConte = useImageEditingContext(),
      isInProgress = _useImageEditingConte.isInProgress,
      zoom = _useImageEditingConte.zoom,
      setZoom = _useImageEditingConte.setZoom;

  return createElement(Dropdown, {
    contentClassName: "wp-block-image__zoom",
    popoverProps: POPOVER_PROPS,
    renderToggle: function renderToggle(_ref) {
      var isOpen = _ref.isOpen,
          onToggle = _ref.onToggle;
      return createElement(ToolbarButton, {
        icon: search,
        label: __('Zoom'),
        onClick: onToggle,
        "aria-expanded": isOpen,
        disabled: isInProgress
      });
    },
    renderContent: function renderContent() {
      return createElement(RangeControl, {
        label: __('Zoom'),
        min: MIN_ZOOM,
        max: MAX_ZOOM,
        value: Math.round(zoom),
        onChange: setZoom
      });
    }
  });
}
