// @wordpress/core-data/build/controls.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.regularFetch = regularFetch;
exports.getDispatch = getDispatch;
exports.default = void 0;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _data = require("@wordpress/data");

/**
 * WordPress dependencies
 */
function regularFetch(url) {
  return {
    type: 'REGULAR_FETCH',
    url: url
  };
}

function getDispatch() {
  return {
    type: 'GET_DISPATCH'
  };
}

var controls = {
  REGULAR_FETCH: function REGULAR_FETCH(_ref) {
    return (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
      var url, _yield$window$fetch$t, data;

      return _regenerator.default.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              url = _ref.url;
              _context.next = 3;
              return window.fetch(url).then(function (res) {
                return res.json();
              });

            case 3:
              _yield$window$fetch$t = _context.sent;
              data = _yield$window$fetch$t.data;
              return _context.abrupt("return", data);

            case 6:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }))();
  },
  GET_DISPATCH: (0, _data.createRegistryControl)(function (_ref2) {
    var dispatch = _ref2.dispatch;
    return function () {
      return dispatch;
    };
  })
};
var _default = controls;
exports.default = _default;
