// @wordpress/block-library/build/navigation/placeholder.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _blocks = require("@wordpress/blocks");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _createDataTree = _interopRequireDefault(require("./create-data-tree"));

var _placeholderPreview = _interopRequireDefault(require("./placeholder-preview"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */

/**
 * A recursive function that maps menu item nodes to blocks.
 *
 * @param {Object[]} menuItems An array of menu items.
 * @return {WPBlock[]} An array of blocks.
 */
function mapMenuItemsToBlocks(menuItems) {
  return menuItems.map(function (menuItem) {
    var _menuItem$xfn, _menuItem$classes, _menuItem$children;

    if (menuItem.type === 'block') {
      var _parse = (0, _blocks.parse)(menuItem.content.raw),
          _parse2 = (0, _slicedToArray2.default)(_parse, 1),
          block = _parse2[0];

      if (!block) {
        return (0, _blocks.createBlock)('core/freeform', {
          content: menuItem.content
        });
      }

      return block;
    }

    var attributes = {
      label: !menuItem.title.rendered ? (0, _i18n.__)('(no title)') : menuItem.title.rendered,
      opensInNewTab: menuItem.target === '_blank'
    };

    if (menuItem.url) {
      attributes.url = menuItem.url;
    }

    if (menuItem.description) {
      attributes.description = menuItem.description;
    }

    if ((_menuItem$xfn = menuItem.xfn) !== null && _menuItem$xfn !== void 0 && _menuItem$xfn.length && (0, _lodash.some)(menuItem.xfn)) {
      attributes.rel = menuItem.xfn.join(' ');
    }

    if ((_menuItem$classes = menuItem.classes) !== null && _menuItem$classes !== void 0 && _menuItem$classes.length && (0, _lodash.some)(menuItem.classes)) {
      attributes.className = menuItem.classes.join(' ');
    }

    var innerBlocks = (_menuItem$children = menuItem.children) !== null && _menuItem$children !== void 0 && _menuItem$children.length ? mapMenuItemsToBlocks(menuItem.children) : [];
    return (0, _blocks.createBlock)('core/navigation-link', attributes, innerBlocks);
  });
}
/**
 * Convert a flat menu item structure to a nested blocks structure.
 *
 * @param {Object[]} menuItems An array of menu items.
 *
 * @return {WPBlock[]} An array of blocks.
 */


function convertMenuItemsToBlocks(menuItems) {
  if (!menuItems) {
    return null;
  }

  var menuTree = (0, _createDataTree.default)(menuItems);
  return mapMenuItemsToBlocks(menuTree);
}
/**
 * Convert pages to blocks.
 *
 * @param {Object[]} pages An array of pages.
 *
 * @return {WPBlock[]} An array of blocks.
 */


function convertPagesToBlocks(pages) {
  if (!pages) {
    return null;
  }

  return pages.map(function (_ref) {
    var title = _ref.title,
        type = _ref.type,
        url = _ref.link,
        id = _ref.id;
    return (0, _blocks.createBlock)('core/navigation-link', {
      type: type,
      id: id,
      url: url,
      label: !title.rendered ? (0, _i18n.__)('(no title)') : title.rendered,
      opensInNewTab: false
    });
  });
}

function NavigationPlaceholder(_ref2, ref) {
  var onCreate = _ref2.onCreate;

  var _useState = (0, _element.useState)(),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      selectedMenu = _useState2[0],
      setSelectedMenu = _useState2[1];

  var _useState3 = (0, _element.useState)(false),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      isCreatingFromMenu = _useState4[0],
      setIsCreatingFromMenu = _useState4[1];

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select('core'),
        getEntityRecords = _select.getEntityRecords,
        getMenus = _select.getMenus,
        getMenuItems = _select.getMenuItems,
        isResolving = _select.isResolving,
        hasFinishedResolution = _select.hasFinishedResolution;

    var pagesParameters = ['postType', 'page', {
      parent: 0,
      order: 'asc',
      orderby: 'id',
      per_page: -1
    }];
    var menusParameters = [{
      per_page: -1
    }];
    var hasSelectedMenu = selectedMenu !== undefined;
    var menuItemsParameters = hasSelectedMenu ? [{
      menus: selectedMenu,
      per_page: -1
    }] : undefined;
    return {
      pages: getEntityRecords.apply(void 0, pagesParameters),
      isResolvingPages: isResolving('getEntityRecords', pagesParameters),
      hasResolvedPages: hasFinishedResolution('getEntityRecords', pagesParameters),
      menus: getMenus.apply(void 0, menusParameters),
      isResolvingMenus: isResolving('getMenus', menusParameters),
      hasResolvedMenus: hasFinishedResolution('getMenus', menusParameters),
      menuItems: hasSelectedMenu ? getMenuItems.apply(void 0, (0, _toConsumableArray2.default)(menuItemsParameters)) : undefined,
      hasResolvedMenuItems: hasSelectedMenu ? hasFinishedResolution('getMenuItems', menuItemsParameters) : false
    };
  }, [selectedMenu]),
      pages = _useSelect.pages,
      isResolvingPages = _useSelect.isResolvingPages,
      hasResolvedPages = _useSelect.hasResolvedPages,
      menus = _useSelect.menus,
      isResolvingMenus = _useSelect.isResolvingMenus,
      hasResolvedMenus = _useSelect.hasResolvedMenus,
      menuItems = _useSelect.menuItems,
      hasResolvedMenuItems = _useSelect.hasResolvedMenuItems;

  var hasPages = !!(hasResolvedPages && pages !== null && pages !== void 0 && pages.length);
  var hasMenus = !!(hasResolvedMenus && menus !== null && menus !== void 0 && menus.length);
  var isLoading = isResolvingPages || isResolvingMenus;
  var createFromMenu = (0, _element.useCallback)(function () {
    var blocks = convertMenuItemsToBlocks(menuItems);
    var selectNavigationBlock = true;
    onCreate(blocks, selectNavigationBlock);
  });

  var onCreateFromMenu = function onCreateFromMenu() {
    // If we have menu items, create the block right away.
    if (hasResolvedMenuItems) {
      createFromMenu();
      return;
    } // Otherwise, create the block when resolution finishes.


    setIsCreatingFromMenu(true);
  };

  var onCreateEmptyMenu = function onCreateEmptyMenu() {
    onCreate([]);
  };

  var onCreateAllPages = function onCreateAllPages() {
    var blocks = convertPagesToBlocks(pages);
    var selectNavigationBlock = true;
    onCreate(blocks, selectNavigationBlock);
  };

  (0, _element.useEffect)(function () {
    // If the user selected a menu but we had to wait for menu items to
    // finish resolving, then create the block once resolution finishes.
    if (isCreatingFromMenu && hasResolvedMenuItems) {
      createFromMenu();
      setIsCreatingFromMenu(false);
    }
  }, [isCreatingFromMenu, hasResolvedMenuItems]);
  return (0, _element.createElement)("div", {
    className: "wp-block-navigation-placeholder"
  }, (0, _element.createElement)(_placeholderPreview.default, null), (0, _element.createElement)("div", {
    className: "wp-block-navigation-placeholder__controls"
  }, isLoading && (0, _element.createElement)("div", {
    ref: ref
  }, (0, _element.createElement)(_components.Spinner, null)), !isLoading && (0, _element.createElement)("div", {
    ref: ref,
    className: "wp-block-navigation-placeholder__actions"
  }, hasMenus ? (0, _element.createElement)(_components.DropdownMenu, {
    text: (0, _i18n.__)('Existing menu'),
    icon: _icons.chevronDown,
    className: "wp-block-navigation-placeholder__actions__dropdown"
  }, function (_ref3) {
    var onClose = _ref3.onClose;
    return (0, _element.createElement)(_components.MenuGroup, null, menus.map(function (menu) {
      return (0, _element.createElement)(_components.MenuItem, {
        onClick: function onClick() {
          setSelectedMenu(menu.id);
          onCreateFromMenu();
        },
        onClose: onClose,
        key: menu.id
      }, menu.name);
    }));
  }) : undefined, hasPages ? (0, _element.createElement)(_components.Button, {
    onClick: onCreateAllPages
  }, (0, _i18n.__)('Add all pages')) : undefined, (0, _element.createElement)(_components.Button, {
    onClick: onCreateEmptyMenu
  }, (0, _i18n.__)('Start empty')))));
}

var _default = (0, _element.forwardRef)(NavigationPlaceholder);

exports.default = _default;
