// @wordpress/block-library/build/image/image-editing/aspect-ratio-dropdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = AspectRatioDropdown;

var _element = require("@wordpress/element");

var _icons = require("@wordpress/icons");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _constants = require("./constants");

var _context = require("./context");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function AspectGroup(_ref) {
  var aspectRatios = _ref.aspectRatios,
      isDisabled = _ref.isDisabled,
      label = _ref.label,
      _onClick = _ref.onClick,
      value = _ref.value;
  return (0, _element.createElement)(_components.MenuGroup, {
    label: label
  }, aspectRatios.map(function (_ref2) {
    var title = _ref2.title,
        aspect = _ref2.aspect;
    return (0, _element.createElement)(_components.MenuItem, {
      key: aspect,
      disabled: isDisabled,
      onClick: function onClick() {
        _onClick(aspect);
      },
      role: "menuitemradio",
      isSelected: aspect === value,
      icon: aspect === value ? _icons.check : undefined
    }, title);
  }));
}

function AspectRatioDropdown(_ref3) {
  var toggleProps = _ref3.toggleProps;

  var _useImageEditingConte = (0, _context.useImageEditingContext)(),
      isInProgress = _useImageEditingConte.isInProgress,
      aspect = _useImageEditingConte.aspect,
      setAspect = _useImageEditingConte.setAspect,
      defaultAspect = _useImageEditingConte.defaultAspect;

  return (0, _element.createElement)(_components.DropdownMenu, {
    icon: _icons.aspectRatio,
    label: (0, _i18n.__)('Aspect Ratio'),
    popoverProps: _constants.POPOVER_PROPS,
    toggleProps: toggleProps,
    className: "wp-block-image__aspect-ratio"
  }, function (_ref4) {
    var onClose = _ref4.onClose;
    return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(AspectGroup, {
      isDisabled: isInProgress,
      onClick: function onClick(newAspect) {
        setAspect(newAspect);
        onClose();
      },
      value: aspect,
      aspectRatios: [{
        title: (0, _i18n.__)('Original'),
        aspect: defaultAspect
      }, {
        title: (0, _i18n.__)('Square'),
        aspect: 1
      }]
    }), (0, _element.createElement)(AspectGroup, {
      label: (0, _i18n.__)('Landscape'),
      isDisabled: isInProgress,
      onClick: function onClick(newAspect) {
        setAspect(newAspect);
        onClose();
      },
      value: aspect,
      aspectRatios: [{
        title: (0, _i18n.__)('16:10'),
        aspect: 16 / 10
      }, {
        title: (0, _i18n.__)('16:9'),
        aspect: 16 / 9
      }, {
        title: (0, _i18n.__)('4:3'),
        aspect: 4 / 3
      }, {
        title: (0, _i18n.__)('3:2'),
        aspect: 3 / 2
      }]
    }), (0, _element.createElement)(AspectGroup, {
      label: (0, _i18n.__)('Portrait'),
      isDisabled: isInProgress,
      onClick: function onClick(newAspect) {
        setAspect(newAspect);
        onClose();
      },
      value: aspect,
      aspectRatios: [{
        title: (0, _i18n.__)('10:16'),
        aspect: 10 / 16
      }, {
        title: (0, _i18n.__)('9:16'),
        aspect: 9 / 16
      }, {
        title: (0, _i18n.__)('3:4'),
        aspect: 3 / 4
      }, {
        title: (0, _i18n.__)('2:3'),
        aspect: 2 / 3
      }]
    }));
  });
}
