// @wordpress/block-library/build/image/image-editing/context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ImageEditingProvider;
exports.useImageEditingContext = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _useSaveImage = _interopRequireDefault(require("./use-save-image"));

var _useTransformImage = _interopRequireDefault(require("./use-transform-image"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var ImageEditingContext = (0, _element.createContext)({});

var useImageEditingContext = function useImageEditingContext() {
  return (0, _element.useContext)(ImageEditingContext);
};

exports.useImageEditingContext = useImageEditingContext;

function ImageEditingProvider(_ref) {
  var id = _ref.id,
      url = _ref.url,
      naturalWidth = _ref.naturalWidth,
      naturalHeight = _ref.naturalHeight,
      isEditing = _ref.isEditing,
      onFinishEditing = _ref.onFinishEditing,
      onSaveImage = _ref.onSaveImage,
      children = _ref.children;
  var transformImage = (0, _useTransformImage.default)({
    url: url,
    naturalWidth: naturalWidth,
    naturalHeight: naturalHeight
  }, isEditing);
  var saveImage = (0, _useSaveImage.default)(_objectSpread({
    id: id,
    url: url,
    onSaveImage: onSaveImage,
    onFinishEditing: onFinishEditing
  }, transformImage));
  var providerValue = (0, _element.useMemo)(function () {
    return _objectSpread(_objectSpread({}, transformImage), saveImage);
  }, [transformImage, saveImage]);
  return (0, _element.createElement)(ImageEditingContext.Provider, {
    value: providerValue
  }, children);
}
