// @wordpress/block-library/build/image/image-editing/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ImageEditor;
Object.defineProperty(exports, "ImageEditingProvider", {
  enumerable: true,
  get: function get() {
    return _context.default;
  }
});

var _element = require("@wordpress/element");

var _blockEditor = require("@wordpress/block-editor");

var _components = require("@wordpress/components");

var _cropper = _interopRequireDefault(require("./cropper"));

var _zoomDropdown = _interopRequireDefault(require("./zoom-dropdown"));

var _aspectRatioDropdown = _interopRequireDefault(require("./aspect-ratio-dropdown"));

var _rotationButton = _interopRequireDefault(require("./rotation-button"));

var _formControls = _interopRequireDefault(require("./form-controls"));

var _context = _interopRequireDefault(require("./context"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ImageEditor(_ref) {
  var url = _ref.url,
      width = _ref.width,
      height = _ref.height,
      clientWidth = _ref.clientWidth,
      naturalHeight = _ref.naturalHeight,
      naturalWidth = _ref.naturalWidth;
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_cropper.default, {
    url: url,
    width: width,
    height: height,
    clientWidth: clientWidth,
    naturalHeight: naturalHeight,
    naturalWidth: naturalWidth
  }), (0, _element.createElement)(_blockEditor.BlockControls, null, (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_zoomDropdown.default, null), (0, _element.createElement)(_components.ToolbarItem, null, function (toggleProps) {
    return (0, _element.createElement)(_aspectRatioDropdown.default, {
      toggleProps: toggleProps
    });
  })), (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_rotationButton.default, null)), (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_formControls.default, null))));
}
