// @wordpress/block-library/build-module/heading/index.js
/**
 * External dependencies
 */
import { isEmpty } from 'lodash';
/**
 * WordPress dependencies
 */

import { heading as icon } from '@wordpress/icons';
import { __, _x, sprintf } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import deprecated from './deprecated';
import edit from './edit';
var metadata = {
  apiVersion: 2,
  name: "core/heading",
  category: "text",
  attributes: {
    textAlign: {
      type: "string"
    },
    content: {
      type: "string",
      source: "html",
      selector: "h1,h2,h3,h4,h5,h6",
      "default": ""
    },
    level: {
      type: "number",
      "default": 2
    },
    placeholder: {
      type: "string"
    }
  },
  supports: {
    align: ["wide", "full"],
    anchor: true,
    className: false,
    color: {
      link: true
    },
    fontSize: true,
    lineHeight: true,
    __experimentalSelector: {
      "core/heading/h1": {
        selector: "h1",
        title: "h1",
        attributes: {
          level: 1
        }
      },
      "core/heading/h2": {
        selector: "h2",
        title: "h2",
        attributes: {
          level: 2
        }
      },
      "core/heading/h3": {
        selector: "h3",
        title: "h3",
        attributes: {
          level: 3
        }
      },
      "core/heading/h4": {
        selector: "h4",
        title: "h4",
        attributes: {
          level: 4
        }
      },
      "core/heading/h5": {
        selector: "h5",
        title: "h5",
        attributes: {
          level: 5
        }
      },
      "core/heading/h6": {
        selector: "h6",
        title: "h6",
        attributes: {
          level: 6
        }
      }
    },
    __unstablePasteTextInline: true
  },
  editorStyle: "wp-block-heading-editor",
  style: "wp-block-heading"
};
import save from './save';
import transforms from './transforms';
var name = metadata.name;
export { metadata, name };
export var settings = {
  title: _x('Heading', 'block title'),
  description: __('Introduce new sections and organize content to help visitors (and search engines) understand the structure of your content.'),
  icon: icon,
  keywords: [__('title'), __('subtitle')],
  example: {
    attributes: {
      content: __('Code is Poetry'),
      level: 2
    }
  },
  __experimentalLabel: function __experimentalLabel(attributes, _ref) {
    var context = _ref.context;

    if (context === 'accessibility') {
      var content = attributes.content,
          level = attributes.level;
      return isEmpty(content) ? sprintf(
      /* translators: accessibility text. %s: heading level. */
      __('Level %s. Empty.'), level) : sprintf(
      /* translators: accessibility text. 1: heading level. 2: heading content. */
      __('Level %1$s. %2$s'), level, content);
    }
  },
  transforms: transforms,
  deprecated: deprecated,
  merge: function merge(attributes, attributesToMerge) {
    return {
      content: (attributes.content || '') + (attributesToMerge.content || '')
    };
  },
  edit: edit,
  save: save
};
