// @wordpress/block-library/build-module/video/tracks-editor.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _objectDestructuringEmpty from "@babel/runtime/helpers/esm/objectDestructuringEmpty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement, Fragment } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * WordPress dependencies
 */
import { __, sprintf } from '@wordpress/i18n';
import { NavigableMenu, MenuItem, FormFileUpload, MenuGroup, ToolbarGroup, ToolbarButton, Dropdown, SVG, Rect, Path, Button, TextControl, SelectControl } from '@wordpress/components';
import { MediaUpload, MediaUploadCheck } from '@wordpress/block-editor';
import { upload, media } from '@wordpress/icons';
import { useSelect } from '@wordpress/data';
import { useState } from '@wordpress/element';
var ALLOWED_TYPES = ['text/vtt'];
var DEFAULT_KIND = 'subtitles';
var KIND_OPTIONS = [{
  label: __('Subtitles'),
  value: 'subtitles'
}, {
  label: __('Captions'),
  value: 'captions'
}, {
  label: __('Descriptions'),
  value: 'descriptions'
}, {
  label: __('Chapters'),
  value: 'chapters'
}, {
  label: __('Metadata'),
  value: 'metadata'
}];
var captionIcon = createElement(SVG, {
  width: "18",
  height: "14",
  viewBox: "0 0 18 14",
  role: "img",
  fill: "none"
}, createElement(Rect, {
  x: "0.75",
  y: "0.75",
  width: "16.5",
  height: "12.5",
  rx: "1.25",
  stroke: "black",
  strokeWidth: "1.5",
  fill: "none"
}), createElement(Path, {
  d: "M3 7H15",
  stroke: "black",
  strokeWidth: "1.5"
}), createElement(Path, {
  d: "M3 10L15 10",
  stroke: "black",
  strokeWidth: "1.5"
}));

function TrackList(_ref) {
  var tracks = _ref.tracks,
      onEditPress = _ref.onEditPress;
  var content;

  if (tracks.length === 0) {
    content = createElement("p", {
      className: "block-library-video-tracks-editor__tracks-informative-message"
    }, __('Tracks can be subtitles, captions, chapters, or descriptions. They help make your content more accessible to a wider range of users.'));
  } else {
    content = tracks.map(function (track, index) {
      return createElement("div", {
        key: index,
        className: "block-library-video-tracks-editor__track-list-track"
      }, createElement("span", null, track.label, " "), createElement(Button, {
        isTertiary: true,
        onClick: function onClick() {
          return onEditPress(index);
        },
        "aria-label": sprintf(
        /* translators: %s: Label of the video text track e.g: "French subtitles" */
        __('Edit %s'), track.label)
      }, __('Edit')));
    });
  }

  return createElement(MenuGroup, {
    label: __('Text tracks'),
    className: "block-library-video-tracks-editor__track-list"
  }, content);
}

function SingleTrackEditor(_ref2) {
  var track = _ref2.track,
      _onChange = _ref2.onChange,
      onClose = _ref2.onClose,
      onRemove = _ref2.onRemove;
  var _track$src = track.src,
      src = _track$src === void 0 ? '' : _track$src,
      _track$label = track.label,
      label = _track$label === void 0 ? '' : _track$label,
      _track$srcLang = track.srcLang,
      srcLang = _track$srcLang === void 0 ? '' : _track$srcLang,
      _track$kind = track.kind,
      kind = _track$kind === void 0 ? DEFAULT_KIND : _track$kind;
  var fileName = src.startsWith('blob:') ? '' : src.substring(src.lastIndexOf('/') + 1);
  return createElement(NavigableMenu, null, createElement("div", {
    className: "block-library-video-tracks-editor__single-track-editor"
  }, createElement("span", {
    className: "block-library-video-tracks-editor__single-track-editor-edit-track-label"
  }, __('Edit track')), createElement("span", null, __('File'), ": ", createElement("b", null, fileName)), createElement("div", {
    className: "block-library-video-tracks-editor__single-track-editor-label-language"
  }, createElement(TextControl
  /* eslint-disable jsx-a11y/no-autofocus */
  , {
    autoFocus: true
    /* eslint-enable jsx-a11y/no-autofocus */
    ,
    onChange: function onChange(newLabel) {
      return _onChange(_objectSpread(_objectSpread({}, track), {}, {
        label: newLabel
      }));
    },
    label: __('Label'),
    value: label,
    help: __('Title of track')
  }), createElement(TextControl, {
    onChange: function onChange(newSrcLang) {
      return _onChange(_objectSpread(_objectSpread({}, track), {}, {
        srcLang: newSrcLang
      }));
    },
    label: __('Source language'),
    value: srcLang,
    help: __('Language tag (en, fr, etc.)')
  })), createElement(SelectControl, {
    className: "block-library-video-tracks-editor__single-track-editor-kind-select",
    options: KIND_OPTIONS,
    value: kind,
    label: __('Kind'),
    onChange: function onChange(newKind) {
      if (newKind === DEFAULT_KIND) {
        newKind = undefined;
      }

      _onChange(_objectSpread(_objectSpread({}, track), {}, {
        kind: newKind
      }));
    }
  }), createElement("div", {
    className: "block-library-video-tracks-editor__single-track-editor-buttons-container"
  }, createElement(Button, {
    isSecondary: true,
    onClick: function onClick() {
      var changes = {};
      var hasChanges = false;

      if (label === '') {
        changes.label = __('English');
        hasChanges = true;
      }

      if (srcLang === '') {
        changes.srcLang = 'en';
        hasChanges = true;
      }

      if (hasChanges) {
        _onChange(_objectSpread(_objectSpread({}, track), changes));
      }

      onClose();
    }
  }, __('Close')), createElement(Button, {
    isDestructive: true,
    isLink: true,
    onClick: onRemove
  }, __('Remove track')))));
}

export default function TracksEditor(_ref3) {
  var _ref3$tracks = _ref3.tracks,
      tracks = _ref3$tracks === void 0 ? [] : _ref3$tracks,
      _onChange2 = _ref3.onChange;
  var mediaUpload = useSelect(function (select) {
    return select('core/block-editor').getSettings().mediaUpload;
  }, []);

  var _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      trackBeingEdited = _useState2[0],
      setTrackBeingEdited = _useState2[1];

  if (!mediaUpload) {
    return null;
  }

  return createElement(Dropdown, {
    contentClassName: "block-library-video-tracks-editor",
    renderToggle: function renderToggle(_ref4) {
      var isOpen = _ref4.isOpen,
          onToggle = _ref4.onToggle;
      return createElement(ToolbarGroup, null, createElement(ToolbarButton, {
        label: __('Text tracks'),
        showTooltip: true,
        "aria-expanded": isOpen,
        "aria-haspopup": "true",
        onClick: onToggle,
        icon: captionIcon
      }));
    },
    renderContent: function renderContent(_ref5) {
      _objectDestructuringEmpty(_ref5);

      if (trackBeingEdited !== null) {
        return createElement(SingleTrackEditor, {
          track: tracks[trackBeingEdited],
          onChange: function onChange(newTrack) {
            var newTracks = _toConsumableArray(tracks);

            newTracks[trackBeingEdited] = newTrack;

            _onChange2(newTracks);
          },
          onClose: function onClose() {
            return setTrackBeingEdited(null);
          },
          onRemove: function onRemove() {
            _onChange2(tracks.filter(function (_track, index) {
              return index !== trackBeingEdited;
            }));

            setTrackBeingEdited(null);
          }
        });
      }

      return createElement(Fragment, null, createElement(NavigableMenu, null, createElement(TrackList, {
        tracks: tracks,
        onEditPress: setTrackBeingEdited
      }), createElement(MenuGroup, {
        className: "block-library-video-tracks-editor__add-tracks-container",
        label: __('Add tracks')
      }, createElement(MediaUpload, {
        onSelect: function onSelect(_ref6) {
          var url = _ref6.url;
          var trackIndex = tracks.length;

          _onChange2([].concat(_toConsumableArray(tracks), [{
            src: url
          }]));

          setTrackBeingEdited(trackIndex);
        },
        allowedTypes: ALLOWED_TYPES,
        render: function render(_ref7) {
          var open = _ref7.open;
          return createElement(MenuItem, {
            icon: media,
            onClick: open
          }, __('Open Media Library'));
        }
      }), createElement(MediaUploadCheck, null, createElement(FormFileUpload, {
        onChange: function onChange(event) {
          var files = event.target.files;
          var trackIndex = tracks.length;
          mediaUpload({
            allowedTypes: ALLOWED_TYPES,
            filesList: files,
            onFileChange: function onFileChange(_ref8) {
              var _ref9 = _slicedToArray(_ref8, 1),
                  url = _ref9[0].url;

              var newTracks = _toConsumableArray(tracks);

              if (!newTracks[trackIndex]) {
                newTracks[trackIndex] = {};
              }

              newTracks[trackIndex] = _objectSpread(_objectSpread({}, tracks[trackIndex]), {}, {
                src: url
              });

              _onChange2(newTracks);

              setTrackBeingEdited(trackIndex);
            }
          });
        },
        accept: ".vtt,text/vtt",
        render: function render(_ref10) {
          var openFileDialog = _ref10.openFileDialog;
          return createElement(MenuItem, {
            icon: upload,
            onClick: function onClick() {
              openFileDialog();
            }
          }, __('Upload'));
        }
      })))));
    }
  });
}
