// @wordpress/block-library/build/column/column-preview.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _reactNative = require("react-native");

var _compose = require("@wordpress/compose");

var _components = require("@wordpress/components");

var _editor = _interopRequireDefault(require("./editor.scss"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
function ColumnPreviewItem(_ref) {
  var index = _ref.index,
      selectedColumnIndex = _ref.selectedColumnIndex,
      width = _ref.width;
  var columnIndicatorStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.columnIndicator, _editor.default.columnIndicatorDark);
  var isSelectedColumn = index === selectedColumnIndex;
  var convertedWidth = (0, _components.useConvertUnitToMobile)(width);
  return (0, _element.createElement)(_reactNative.View, {
    style: [isSelectedColumn && columnIndicatorStyle, {
      flex: convertedWidth
    }],
    key: index
  });
}

function ColumnsPreview(_ref2) {
  var columnWidths = _ref2.columnWidths,
      selectedColumnIndex = _ref2.selectedColumnIndex;
  var columnsPreviewStyle = (0, _compose.usePreferredColorSchemeStyle)(_editor.default.columnsPreview, _editor.default.columnsPreviewDark);
  return (0, _element.createElement)(_reactNative.View, {
    style: columnsPreviewStyle
  }, columnWidths.map(function (width, index) {
    return (0, _element.createElement)(ColumnPreviewItem, {
      index: index,
      selectedColumnIndex: selectedColumnIndex,
      width: width,
      key: index
    });
  }));
}

var _default = ColumnsPreview;
exports.default = _default;
