// @wordpress/block-library/build/column/edit.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _reactNative = require("react-native");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

var _blockEditor = require("@wordpress/block-editor");

var _components = require("@wordpress/components");

var _i18n = require("@wordpress/i18n");

var _editor = _interopRequireDefault(require("./editor.scss"));

var _columnPreview = _interopRequireDefault(require("./column-preview"));

var _utils = require("../columns/utils");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

var isWider = _components.alignmentHelpers.isWider;

function ColumnEdit(_ref) {
  var attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      hasChildren = _ref.hasChildren,
      isSelected = _ref.isSelected,
      getStylesFromColorScheme = _ref.getStylesFromColorScheme,
      isParentSelected = _ref.isParentSelected,
      contentStyle = _ref.contentStyle,
      columns = _ref.columns,
      selectedColumnIndex = _ref.selectedColumnIndex,
      parentAlignment = _ref.parentAlignment,
      clientId = _ref.clientId,
      blockWidth = _ref.blockWidth;
  var verticalAlignment = attributes.verticalAlignment,
      width = attributes.width;

  var _ref2 = (0, _components.getValueAndUnit)(width) || {},
      _ref2$valueUnit = _ref2.valueUnit,
      valueUnit = _ref2$valueUnit === void 0 ? '%' : _ref2$valueUnit;

  var screenWidth = Math.floor(_reactNative.Dimensions.get('window').width);

  var _useState = (0, _element.useState)(valueUnit || '%'),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      widthUnit = _useState2[0],
      setWidthUnit = _useState2[1];

  var updateAlignment = function updateAlignment(alignment) {
    setAttributes({
      verticalAlignment: alignment
    });
  };

  (0, _element.useEffect)(function () {
    setWidthUnit(valueUnit);
  }, [valueUnit]);
  (0, _element.useEffect)(function () {
    if (!verticalAlignment && parentAlignment) {
      updateAlignment(parentAlignment);
    }
  }, []);

  var onChangeWidth = function onChangeWidth(nextWidth) {
    var widthWithUnit = (0, _utils.getWidthWithUnit)(nextWidth, widthUnit);
    setAttributes({
      width: widthWithUnit
    });
  };

  var onChangeUnit = function onChangeUnit(nextUnit) {
    setWidthUnit(nextUnit);
    var widthWithoutUnit = parseFloat(width || (0, _utils.getWidths)(columns)[selectedColumnIndex]);
    setAttributes({
      width: (0, _utils.getWidthWithUnit)(widthWithoutUnit, nextUnit)
    });
  };

  var onChange = function onChange(nextWidth) {
    if ((0, _utils.isPercentageUnit)(widthUnit) || !widthUnit) {
      return;
    }

    onChangeWidth(nextWidth);
  };

  var renderAppender = (0, _element.useCallback)(function () {
    var columnWidth = contentStyle[clientId].width;
    var isScreenWidthEqual = columnWidth === screenWidth;

    if (isSelected) {
      return (0, _element.createElement)(_reactNative.View, {
        style: (isWider(screenWidth, 'mobile') || isScreenWidthEqual) && (hasChildren ? _editor.default.columnAppender : _editor.default.wideColumnAppender)
      }, (0, _element.createElement)(_blockEditor.InnerBlocks.ButtonBlockAppender, null));
    }

    return null;
  }, [contentStyle[clientId], screenWidth, isSelected, hasChildren]);

  if (!isSelected && !hasChildren) {
    return (0, _element.createElement)(_reactNative.View, {
      style: [!isParentSelected && getStylesFromColorScheme(_editor.default.columnPlaceholder, _editor.default.columnPlaceholderDark), _editor.default.columnPlaceholderNotSelected, contentStyle[clientId]]
    });
  }

  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.BlockControls, null, (0, _element.createElement)(_blockEditor.BlockVerticalAlignmentToolbar, {
    onChange: updateAlignment,
    value: verticalAlignment
  })), (0, _element.createElement)(_blockEditor.InspectorControls, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Column settings')
  }, (0, _element.createElement)(_components.UnitControl, {
    label: (0, _i18n.__)('Width'),
    min: 1,
    max: (0, _utils.isPercentageUnit)(widthUnit) ? 100 : undefined,
    onChange: onChange,
    onComplete: onChangeWidth,
    onUnitChange: onChangeUnit,
    decimalNum: 1,
    value: (0, _utils.getWidths)(columns)[selectedColumnIndex],
    unit: widthUnit,
    units: _utils.CSS_UNITS,
    preview: (0, _element.createElement)(_columnPreview.default, {
      columnWidths: (0, _utils.getWidths)(columns, false),
      selectedColumnIndex: selectedColumnIndex
    })
  })), (0, _element.createElement)(_components.PanelBody, null, (0, _element.createElement)(_components.FooterMessageControl, {
    label: (0, _i18n.__)('Note: Column layout may vary between themes and screen sizes')
  }))), (0, _element.createElement)(_reactNative.View, {
    style: [isSelected && hasChildren && _editor.default.innerBlocksBottomSpace, contentStyle[clientId]]
  }, (0, _element.createElement)(_blockEditor.InnerBlocks, {
    renderAppender: renderAppender,
    parentWidth: contentStyle[clientId].width,
    blockWidth: blockWidth
  })));
}

function ColumnEditWrapper(props) {
  var verticalAlignment = props.attributes.verticalAlignment;

  var getVerticalAlignmentRemap = function getVerticalAlignmentRemap(alignment) {
    if (!alignment) return _editor.default.flexBase;
    return _objectSpread(_objectSpread({}, _editor.default.flexBase), _editor.default["is-vertically-aligned-".concat(alignment)]);
  };

  return (0, _element.createElement)(_reactNative.View, {
    style: getVerticalAlignmentRemap(verticalAlignment)
  }, (0, _element.createElement)(ColumnEdit, props));
}

var _default = (0, _compose.compose)([(0, _data.withSelect)(function (select, _ref3) {
  var _getBlockAttributes;

  var clientId = _ref3.clientId;

  var _select = select('core/block-editor'),
      getBlockCount = _select.getBlockCount,
      getBlockRootClientId = _select.getBlockRootClientId,
      getSelectedBlockClientId = _select.getSelectedBlockClientId,
      getBlocks = _select.getBlocks,
      getBlockOrder = _select.getBlockOrder,
      getBlockAttributes = _select.getBlockAttributes;

  var selectedBlockClientId = getSelectedBlockClientId();
  var isSelected = selectedBlockClientId === clientId;
  var parentId = getBlockRootClientId(clientId);
  var hasChildren = !!getBlockCount(clientId);
  var isParentSelected = selectedBlockClientId && selectedBlockClientId === parentId;
  var blockOrder = getBlockOrder(parentId);
  var selectedColumnIndex = blockOrder.indexOf(clientId);
  var columns = getBlocks(parentId);
  var parentAlignment = (_getBlockAttributes = getBlockAttributes(parentId)) === null || _getBlockAttributes === void 0 ? void 0 : _getBlockAttributes.verticalAlignment;
  return {
    hasChildren: hasChildren,
    isParentSelected: isParentSelected,
    isSelected: isSelected,
    selectedColumnIndex: selectedColumnIndex,
    columns: columns,
    parentAlignment: parentAlignment
  };
}), _compose.withPreferredColorScheme])(ColumnEditWrapper);

exports.default = _default;
