// @wordpress/block-library/build/query/variations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _icons2 = require("./icons");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var variations = [{
  name: 'posts-list',
  title: (0, _i18n.__)('Posts List'),
  description: (0, _i18n.__)('Display a list of your most recent posts, excluding sticky posts.'),
  icon: _icons.postList,
  attributes: {
    query: {
      perPage: 4,
      pages: 1,
      offset: 0,
      postType: 'post',
      categoryIds: [],
      tagIds: [],
      order: 'desc',
      orderBy: 'date',
      author: '',
      search: '',
      sticky: 'exclude'
    }
  },
  scope: ['inserter']
}, {
  name: 'title-date',
  title: (0, _i18n.__)('Title & Date'),
  icon: _icons2.titleDate,
  innerBlocks: [['core/query-loop', {}, [['core/post-title'], ['core/post-date']]]],
  scope: ['block']
}, {
  name: 'title-excerpt',
  title: (0, _i18n.__)('Title & Excerpt'),
  icon: _icons2.titleExcerpt,
  innerBlocks: [['core/query-loop', {}, [['core/post-title'], ['core/post-excerpt']]]],
  scope: ['block']
}, {
  name: 'title-date-excerpt',
  title: (0, _i18n.__)('Title, Date, & Excerpt'),
  icon: _icons2.titleDateExcerpt,
  innerBlocks: [['core/query-loop', {}, [['core/post-title'], ['core/post-date'], ['core/post-excerpt']]]],
  scope: ['block']
}, {
  name: 'image-date-title',
  title: (0, _i18n.__)('Image, Date, & Title'),
  icon: _icons2.imageDateTitle,
  innerBlocks: [['core/query-loop', {}, [['core/post-featured-image'], ['core/post-date'], ['core/post-title']]]],
  scope: ['block']
}];
var _default = variations;
exports.default = _default;
