// @wordpress/block-library/build/site-logo/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _icon = _interopRequireDefault(require("./icon"));

var _edit = _interopRequireDefault(require("./edit"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/site-logo",
  category: "layout",
  attributes: {
    align: {
      type: "string"
    },
    width: {
      type: "number"
    }
  },
  supports: {
    html: false
  },
  editorStyle: "wp-block-site-logo-editor",
  style: "wp-block-site-logo"
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Site Logo', 'block title'),
  description: (0, _i18n.__)('Show a site logo'),
  icon: _icon.default,
  styles: [{
    name: 'default',
    label: (0, _i18n._x)('Default', 'block style'),
    isDefault: true
  }, {
    name: 'rounded',
    label: (0, _i18n._x)('Rounded', 'block style')
  }],
  supports: {
    align: true,
    alignWide: false
  },
  edit: _edit.default
};
exports.settings = settings;
