// @wordpress/block-library/build-module/audio/edit.native.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { Text, TouchableWithoutFeedback } from 'react-native';
import { isEmpty } from 'lodash';
/**
 * WordPress dependencies
 */

import { View } from '@wordpress/primitives';
import { PanelBody, SelectControl, ToggleControl, withNotices, ToolbarButton, ToolbarGroup } from '@wordpress/components';
import { BlockCaption, BlockControls, BlockIcon, InspectorControls, MediaPlaceholder, MediaUpload, MediaUploadProgress } from '@wordpress/block-editor';
import { __, sprintf } from '@wordpress/i18n';
import { audio as icon, replace } from '@wordpress/icons';
import { useState } from '@wordpress/element';
var ALLOWED_MEDIA_TYPES = ['audio'];

function AudioEdit(_ref) {
  var attributes = _ref.attributes,
      noticeOperations = _ref.noticeOperations,
      setAttributes = _ref.setAttributes,
      isSelected = _ref.isSelected,
      noticeUI = _ref.noticeUI,
      insertBlocksAfter = _ref.insertBlocksAfter,
      onFocus = _ref.onFocus,
      onBlur = _ref.onBlur,
      clientId = _ref.clientId;
  var id = attributes.id,
      autoplay = attributes.autoplay,
      loop = attributes.loop,
      preload = attributes.preload,
      src = attributes.src;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      isCaptionSelected = _useState2[0],
      setIsCaptionSelected = _useState2[1];

  var onFileChange = function onFileChange(_ref2) {
    var mediaId = _ref2.mediaId,
        mediaUrl = _ref2.mediaUrl;
    setAttributes({
      id: mediaId,
      src: mediaUrl
    });
  };

  var onError = function onError() {
    // TODO: Set up error state
    onUploadError(__('Error'));
  };

  function toggleAttribute(attribute) {
    return function (newValue) {
      setAttributes(_defineProperty({}, attribute, newValue));
    };
  }

  function onSelectURL() {// TODO: Set up add audio from URL flow
  }

  function onUploadError(message) {
    noticeOperations.removeAllNotices();
    noticeOperations.createErrorNotice(message);
  } // const { setAttributes, isSelected, noticeUI } = this.props;


  function onSelectAudio(media) {
    if (!media || !media.url) {
      // in this case there was an error and we should continue in the editing state
      // previous attributes should be removed because they may be temporary blob urls
      setAttributes({
        src: undefined,
        id: undefined
      });
      return;
    } // sets the block's attribute and updates the edit component from the
    // selected media, then switches off the editing UI


    setAttributes({
      src: media.url,
      id: media.id
    });
  }

  function onAudioPress() {
    setIsCaptionSelected(false);
  }

  function onFocusCaption() {
    if (!isCaptionSelected) {
      setIsCaptionSelected(true);
    }
  }

  if (!src) {
    return createElement(View, null, createElement(MediaPlaceholder, {
      icon: createElement(BlockIcon, {
        icon: icon
      }),
      onSelect: onSelectAudio,
      onSelectURL: onSelectURL,
      accept: "audio/*",
      allowedTypes: ALLOWED_MEDIA_TYPES,
      value: attributes,
      notices: noticeUI,
      onError: onUploadError,
      onFocus: onFocus
    }));
  }

  function getBlockControls(open) {
    return createElement(BlockControls, null, createElement(ToolbarGroup, null, createElement(ToolbarButton, {
      title: __('Replace audio'),
      icon: replace,
      onClick: open
    })));
  }

  function getBlockUI(open, getMediaOptions) {
    return createElement(MediaUploadProgress, {
      mediaId: id,
      onUpdateMediaProgress: this.updateMediaProgress,
      onFinishMediaUploadWithSuccess: onFileChange,
      onFinishMediaUploadWithFailure: onError,
      onMediaUploadStateReset: onFileChange,
      renderContent: function renderContent(_ref3) {
        var isUploadInProgress = _ref3.isUploadInProgress,
            isUploadFailed = _ref3.isUploadFailed;
        return createElement(View, null, !isCaptionSelected && getBlockControls(open), getMediaOptions(), createElement(Text, null, "\u23EF Audio Player goes here.", ' ', isUploadInProgress && 'Uploading...', isUploadFailed && 'ERROR'));
      }
    });
  }

  return createElement(TouchableWithoutFeedback, {
    accessible: !isSelected,
    onPress: onAudioPress,
    disabled: !isSelected
  }, createElement(View, null, createElement(InspectorControls, null, createElement(PanelBody, {
    title: __('Audio settings')
  }, createElement(ToggleControl, {
    label: __('Autoplay'),
    onChange: toggleAttribute('autoplay'),
    checked: autoplay
  }), createElement(ToggleControl, {
    label: __('Loop'),
    onChange: toggleAttribute('loop'),
    checked: loop
  }), createElement(SelectControl, {
    label: __('Preload'),
    value: preload || '' // `undefined` is required for the preload attribute to be unset.
    ,
    onChange: function onChange(value) {
      return setAttributes({
        preload: value || undefined
      });
    },
    options: [{
      value: '',
      label: __('Browser default')
    }, {
      value: 'auto',
      label: __('Auto')
    }, {
      value: 'metadata',
      label: __('Metadata')
    }, {
      value: 'none',
      label: __('None')
    }]
  }))), createElement(MediaUpload, {
    allowedTypes: ALLOWED_MEDIA_TYPES,
    isReplacingMedia: true,
    onSelect: onSelectAudio,
    render: function render(_ref4) {
      var open = _ref4.open,
          getMediaOptions = _ref4.getMediaOptions;
      return getBlockUI(open, getMediaOptions);
    }
  }), createElement(BlockCaption, {
    accessible: true,
    accessibilityLabelCreator: function accessibilityLabelCreator(caption) {
      return isEmpty(caption) ?
      /* translators: accessibility text. Empty Audio caption. */
      __('Audio caption. Empty') : sprintf(
      /* translators: accessibility text. %s: Audio caption. */
      __('Audio caption. %s'), caption);
    },
    clientId: clientId,
    isSelected: isCaptionSelected,
    onFocus: onFocusCaption,
    onBlur: onBlur,
    insertBlocksAfter: insertBlocksAfter
  })));
}

export default withNotices(AudioEdit);
