// @wordpress/block-library/build-module/audio/index.js
/**
 * WordPress dependencies
 */
import { __, _x } from '@wordpress/i18n';
import { audio as icon } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import deprecated from './deprecated';
import edit from './edit';
var metadata = {
  apiVersion: 2,
  name: "core/audio",
  category: "media",
  attributes: {
    src: {
      type: "string",
      source: "attribute",
      selector: "audio",
      attribute: "src"
    },
    caption: {
      type: "string",
      source: "html",
      selector: "figcaption"
    },
    id: {
      type: "number"
    },
    autoplay: {
      type: "boolean",
      source: "attribute",
      selector: "audio",
      attribute: "autoplay"
    },
    loop: {
      type: "boolean",
      source: "attribute",
      selector: "audio",
      attribute: "loop"
    },
    preload: {
      type: "string",
      source: "attribute",
      selector: "audio",
      attribute: "preload"
    }
  },
  supports: {
    anchor: true,
    align: true
  },
  editorStyle: "wp-block-audio-editor",
  style: "wp-block-audio"
};
import save from './save';
import transforms from './transforms';
var name = metadata.name;
export { metadata, name };
export var settings = {
  title: _x('Audio', 'block title'),
  description: __('Embed a simple audio player.'),
  keywords: [__('music'), __('sound'), __('podcast'), __('recording')],
  icon: icon,
  transforms: transforms,
  deprecated: deprecated,
  edit: edit,
  save: save
};
