// @wordpress/block-library/build/video/edit-common-settings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

/**
 * WordPress dependencies
 */
var options = [{
  value: 'auto',
  label: (0, _i18n.__)('Auto')
}, {
  value: 'metadata',
  label: (0, _i18n.__)('Metadata')
}, {
  value: 'none',
  label: (0, _i18n.__)('None')
}];

var VideoSettings = function VideoSettings(_ref) {
  var setAttributes = _ref.setAttributes,
      attributes = _ref.attributes;
  var autoplay = attributes.autoplay,
      controls = attributes.controls,
      loop = attributes.loop,
      muted = attributes.muted,
      playsInline = attributes.playsInline,
      preload = attributes.preload;
  var getAutoplayHelp = (0, _element.useCallback)(function (checked) {
    return checked ? (0, _i18n.__)('Note: Autoplaying videos may cause usability issues for some visitors.') : null;
  }, []);
  var toggleFactory = (0, _element.useMemo)(function () {
    var toggleAttribute = function toggleAttribute(attribute) {
      return function (newValue) {
        setAttributes((0, _defineProperty2.default)({}, attribute, newValue));
      };
    };

    return {
      autoplay: toggleAttribute('autoplay'),
      loop: toggleAttribute('loop'),
      muted: toggleAttribute('muted'),
      controls: toggleAttribute('controls'),
      playsInline: toggleAttribute('playsInline')
    };
  }, []);
  var onChangePreload = (0, _element.useCallback)(function (value) {
    setAttributes({
      preload: value
    });
  }, []);
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Autoplay'),
    onChange: toggleFactory.autoplay,
    checked: autoplay,
    help: getAutoplayHelp
  }), (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Loop'),
    onChange: toggleFactory.loop,
    checked: loop
  }), (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Muted'),
    onChange: toggleFactory.muted,
    checked: muted
  }), (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Playback controls'),
    onChange: toggleFactory.controls,
    checked: controls
  }), (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Play inline'),
    onChange: toggleFactory.playsInline,
    checked: playsInline
  }), (0, _element.createElement)(_components.SelectControl, {
    label: (0, _i18n.__)('Preload'),
    value: preload,
    onChange: onChangePreload,
    options: options
  }));
};

var _default = VideoSettings;
exports.default = _default;
