// @wordpress/block-library/build-module/column/column-preview.native.js
import { createElement } from "@wordpress/element";

/**
 * External dependencies
 */
import { View } from 'react-native';
/**
 * WordPress dependencies
 */

import { usePreferredColorSchemeStyle } from '@wordpress/compose';
import { useConvertUnitToMobile } from '@wordpress/components';
/**
 * Internal dependencies
 */

import styles from './editor.scss';

function ColumnPreviewItem(_ref) {
  var index = _ref.index,
      selectedColumnIndex = _ref.selectedColumnIndex,
      width = _ref.width;
  var columnIndicatorStyle = usePreferredColorSchemeStyle(styles.columnIndicator, styles.columnIndicatorDark);
  var isSelectedColumn = index === selectedColumnIndex;
  var convertedWidth = useConvertUnitToMobile(width);
  return createElement(View, {
    style: [isSelectedColumn && columnIndicatorStyle, {
      flex: convertedWidth
    }],
    key: index
  });
}

function ColumnsPreview(_ref2) {
  var columnWidths = _ref2.columnWidths,
      selectedColumnIndex = _ref2.selectedColumnIndex;
  var columnsPreviewStyle = usePreferredColorSchemeStyle(styles.columnsPreview, styles.columnsPreviewDark);
  return createElement(View, {
    style: columnsPreviewStyle
  }, columnWidths.map(function (width, index) {
    return createElement(ColumnPreviewItem, {
      index: index,
      selectedColumnIndex: selectedColumnIndex,
      width: width,
      key: index
    });
  }));
}

export default ColumnsPreview;
