// @wordpress/block-library/build-module/column/edit.native.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement, Fragment } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * External dependencies
 */
import { View, Dimensions } from 'react-native';
/**
 * WordPress dependencies
 */

import { withSelect } from '@wordpress/data';
import { compose, withPreferredColorScheme } from '@wordpress/compose';
import { useEffect, useState, useCallback } from '@wordpress/element';
import { InnerBlocks, BlockControls, BlockVerticalAlignmentToolbar, InspectorControls } from '@wordpress/block-editor';
import { PanelBody, FooterMessageControl, UnitControl, getValueAndUnit, alignmentHelpers } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

import styles from './editor.scss';
import ColumnsPreview from './column-preview';
import { getWidths, getWidthWithUnit, isPercentageUnit, CSS_UNITS } from '../columns/utils';
var isWider = alignmentHelpers.isWider;

function ColumnEdit(_ref) {
  var attributes = _ref.attributes,
      setAttributes = _ref.setAttributes,
      hasChildren = _ref.hasChildren,
      isSelected = _ref.isSelected,
      getStylesFromColorScheme = _ref.getStylesFromColorScheme,
      isParentSelected = _ref.isParentSelected,
      contentStyle = _ref.contentStyle,
      columns = _ref.columns,
      selectedColumnIndex = _ref.selectedColumnIndex,
      parentAlignment = _ref.parentAlignment,
      clientId = _ref.clientId,
      blockWidth = _ref.blockWidth;
  var verticalAlignment = attributes.verticalAlignment,
      width = attributes.width;

  var _ref2 = getValueAndUnit(width) || {},
      _ref2$valueUnit = _ref2.valueUnit,
      valueUnit = _ref2$valueUnit === void 0 ? '%' : _ref2$valueUnit;

  var screenWidth = Math.floor(Dimensions.get('window').width);

  var _useState = useState(valueUnit || '%'),
      _useState2 = _slicedToArray(_useState, 2),
      widthUnit = _useState2[0],
      setWidthUnit = _useState2[1];

  var updateAlignment = function updateAlignment(alignment) {
    setAttributes({
      verticalAlignment: alignment
    });
  };

  useEffect(function () {
    setWidthUnit(valueUnit);
  }, [valueUnit]);
  useEffect(function () {
    if (!verticalAlignment && parentAlignment) {
      updateAlignment(parentAlignment);
    }
  }, []);

  var onChangeWidth = function onChangeWidth(nextWidth) {
    var widthWithUnit = getWidthWithUnit(nextWidth, widthUnit);
    setAttributes({
      width: widthWithUnit
    });
  };

  var onChangeUnit = function onChangeUnit(nextUnit) {
    setWidthUnit(nextUnit);
    var widthWithoutUnit = parseFloat(width || getWidths(columns)[selectedColumnIndex]);
    setAttributes({
      width: getWidthWithUnit(widthWithoutUnit, nextUnit)
    });
  };

  var onChange = function onChange(nextWidth) {
    if (isPercentageUnit(widthUnit) || !widthUnit) {
      return;
    }

    onChangeWidth(nextWidth);
  };

  var renderAppender = useCallback(function () {
    var columnWidth = contentStyle[clientId].width;
    var isScreenWidthEqual = columnWidth === screenWidth;

    if (isSelected) {
      return createElement(View, {
        style: (isWider(screenWidth, 'mobile') || isScreenWidthEqual) && (hasChildren ? styles.columnAppender : styles.wideColumnAppender)
      }, createElement(InnerBlocks.ButtonBlockAppender, null));
    }

    return null;
  }, [contentStyle[clientId], screenWidth, isSelected, hasChildren]);

  if (!isSelected && !hasChildren) {
    return createElement(View, {
      style: [!isParentSelected && getStylesFromColorScheme(styles.columnPlaceholder, styles.columnPlaceholderDark), styles.columnPlaceholderNotSelected, contentStyle[clientId]]
    });
  }

  return createElement(Fragment, null, createElement(BlockControls, null, createElement(BlockVerticalAlignmentToolbar, {
    onChange: updateAlignment,
    value: verticalAlignment
  })), createElement(InspectorControls, null, createElement(PanelBody, {
    title: __('Column settings')
  }, createElement(UnitControl, {
    label: __('Width'),
    min: 1,
    max: isPercentageUnit(widthUnit) ? 100 : undefined,
    onChange: onChange,
    onComplete: onChangeWidth,
    onUnitChange: onChangeUnit,
    decimalNum: 1,
    value: getWidths(columns)[selectedColumnIndex],
    unit: widthUnit,
    units: CSS_UNITS,
    preview: createElement(ColumnsPreview, {
      columnWidths: getWidths(columns, false),
      selectedColumnIndex: selectedColumnIndex
    })
  })), createElement(PanelBody, null, createElement(FooterMessageControl, {
    label: __('Note: Column layout may vary between themes and screen sizes')
  }))), createElement(View, {
    style: [isSelected && hasChildren && styles.innerBlocksBottomSpace, contentStyle[clientId]]
  }, createElement(InnerBlocks, {
    renderAppender: renderAppender,
    parentWidth: contentStyle[clientId].width,
    blockWidth: blockWidth
  })));
}

function ColumnEditWrapper(props) {
  var verticalAlignment = props.attributes.verticalAlignment;

  var getVerticalAlignmentRemap = function getVerticalAlignmentRemap(alignment) {
    if (!alignment) return styles.flexBase;
    return _objectSpread(_objectSpread({}, styles.flexBase), styles["is-vertically-aligned-".concat(alignment)]);
  };

  return createElement(View, {
    style: getVerticalAlignmentRemap(verticalAlignment)
  }, createElement(ColumnEdit, props));
}

export default compose([withSelect(function (select, _ref3) {
  var _getBlockAttributes;

  var clientId = _ref3.clientId;

  var _select = select('core/block-editor'),
      getBlockCount = _select.getBlockCount,
      getBlockRootClientId = _select.getBlockRootClientId,
      getSelectedBlockClientId = _select.getSelectedBlockClientId,
      getBlocks = _select.getBlocks,
      getBlockOrder = _select.getBlockOrder,
      getBlockAttributes = _select.getBlockAttributes;

  var selectedBlockClientId = getSelectedBlockClientId();
  var isSelected = selectedBlockClientId === clientId;
  var parentId = getBlockRootClientId(clientId);
  var hasChildren = !!getBlockCount(clientId);
  var isParentSelected = selectedBlockClientId && selectedBlockClientId === parentId;
  var blockOrder = getBlockOrder(parentId);
  var selectedColumnIndex = blockOrder.indexOf(clientId);
  var columns = getBlocks(parentId);
  var parentAlignment = (_getBlockAttributes = getBlockAttributes(parentId)) === null || _getBlockAttributes === void 0 ? void 0 : _getBlockAttributes.verticalAlignment;
  return {
    hasChildren: hasChildren,
    isParentSelected: isParentSelected,
    isSelected: isSelected,
    selectedColumnIndex: selectedColumnIndex,
    columns: columns,
    parentAlignment: parentAlignment
  };
}), withPreferredColorScheme])(ColumnEditWrapper);
