// @wordpress/block-library/build/heading/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _lodash = require("lodash");

var _icons = require("@wordpress/icons");

var _i18n = require("@wordpress/i18n");

var _deprecated = _interopRequireDefault(require("./deprecated"));

var _edit = _interopRequireDefault(require("./edit"));

var _save = _interopRequireDefault(require("./save"));

var _transforms = _interopRequireDefault(require("./transforms"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/heading",
  category: "text",
  attributes: {
    textAlign: {
      type: "string"
    },
    content: {
      type: "string",
      source: "html",
      selector: "h1,h2,h3,h4,h5,h6",
      "default": ""
    },
    level: {
      type: "number",
      "default": 2
    },
    placeholder: {
      type: "string"
    }
  },
  supports: {
    align: ["wide", "full"],
    anchor: true,
    className: false,
    color: {
      link: true
    },
    fontSize: true,
    lineHeight: true,
    __experimentalSelector: {
      "core/heading/h1": {
        selector: "h1",
        title: "h1",
        attributes: {
          level: 1
        }
      },
      "core/heading/h2": {
        selector: "h2",
        title: "h2",
        attributes: {
          level: 2
        }
      },
      "core/heading/h3": {
        selector: "h3",
        title: "h3",
        attributes: {
          level: 3
        }
      },
      "core/heading/h4": {
        selector: "h4",
        title: "h4",
        attributes: {
          level: 4
        }
      },
      "core/heading/h5": {
        selector: "h5",
        title: "h5",
        attributes: {
          level: 5
        }
      },
      "core/heading/h6": {
        selector: "h6",
        title: "h6",
        attributes: {
          level: 6
        }
      }
    },
    __unstablePasteTextInline: true
  },
  editorStyle: "wp-block-heading-editor",
  style: "wp-block-heading"
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Heading', 'block title'),
  description: (0, _i18n.__)('Introduce new sections and organize content to help visitors (and search engines) understand the structure of your content.'),
  icon: _icons.heading,
  keywords: [(0, _i18n.__)('title'), (0, _i18n.__)('subtitle')],
  example: {
    attributes: {
      content: (0, _i18n.__)('Code is Poetry'),
      level: 2
    }
  },
  __experimentalLabel: function __experimentalLabel(attributes, _ref) {
    var context = _ref.context;

    if (context === 'accessibility') {
      var content = attributes.content,
          level = attributes.level;
      return (0, _lodash.isEmpty)(content) ? (0, _i18n.sprintf)(
      /* translators: accessibility text. %s: heading level. */
      (0, _i18n.__)('Level %s. Empty.'), level) : (0, _i18n.sprintf)(
      /* translators: accessibility text. 1: heading level. 2: heading content. */
      (0, _i18n.__)('Level %1$s. %2$s'), level, content);
    }
  },
  transforms: _transforms.default,
  deprecated: _deprecated.default,
  merge: function merge(attributes, attributesToMerge) {
    return {
      content: (attributes.content || '') + (attributesToMerge.content || '')
    };
  },
  edit: _edit.default,
  save: _save.default
};
exports.settings = settings;
