// @wordpress/block-library/build/heading/transforms.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _blocks = require("@wordpress/blocks");

var _shared = require("./shared");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var _apiVersion$name$cate = {
  apiVersion: 2,
  name: "core/heading",
  category: "text",
  attributes: {
    textAlign: {
      type: "string"
    },
    content: {
      type: "string",
      source: "html",
      selector: "h1,h2,h3,h4,h5,h6",
      "default": ""
    },
    level: {
      type: "number",
      "default": 2
    },
    placeholder: {
      type: "string"
    }
  },
  supports: {
    align: ["wide", "full"],
    anchor: true,
    className: false,
    color: {
      link: true
    },
    fontSize: true,
    lineHeight: true,
    __experimentalSelector: {
      "core/heading/h1": {
        selector: "h1",
        title: "h1",
        attributes: {
          level: 1
        }
      },
      "core/heading/h2": {
        selector: "h2",
        title: "h2",
        attributes: {
          level: 2
        }
      },
      "core/heading/h3": {
        selector: "h3",
        title: "h3",
        attributes: {
          level: 3
        }
      },
      "core/heading/h4": {
        selector: "h4",
        title: "h4",
        attributes: {
          level: 4
        }
      },
      "core/heading/h5": {
        selector: "h5",
        title: "h5",
        attributes: {
          level: 5
        }
      },
      "core/heading/h6": {
        selector: "h6",
        title: "h6",
        attributes: {
          level: 6
        }
      }
    },
    __unstablePasteTextInline: true
  },
  editorStyle: "wp-block-heading-editor",
  style: "wp-block-heading"
},
    name = _apiVersion$name$cate.name;
var transforms = {
  from: [{
    type: 'block',
    isMultiBlock: true,
    blocks: ['core/paragraph'],
    transform: function transform(attributes) {
      return attributes.map(function (_ref) {
        var content = _ref.content,
            anchor = _ref.anchor;
        return (0, _blocks.createBlock)(name, {
          content: content,
          anchor: anchor
        });
      });
    }
  }, {
    type: 'raw',
    selector: 'h1,h2,h3,h4,h5,h6',
    schema: function schema(_ref2) {
      var phrasingContentSchema = _ref2.phrasingContentSchema,
          isPaste = _ref2.isPaste;
      var schema = {
        children: phrasingContentSchema,
        attributes: isPaste ? [] : ['style', 'id']
      };
      return {
        h1: schema,
        h2: schema,
        h3: schema,
        h4: schema,
        h5: schema,
        h6: schema
      };
    },
    transform: function transform(node) {
      var attributes = (0, _blocks.getBlockAttributes)(name, node.outerHTML);

      var _ref3 = node.style || {},
          textAlign = _ref3.textAlign;

      attributes.level = (0, _shared.getLevelFromHeadingNodeName)(node.nodeName);

      if (textAlign === 'left' || textAlign === 'center' || textAlign === 'right') {
        attributes.align = textAlign;
      }

      return (0, _blocks.createBlock)(name, attributes);
    }
  }].concat((0, _toConsumableArray2.default)([1, 2, 3, 4, 5, 6].map(function (level) {
    return {
      type: 'prefix',
      prefix: Array(level + 1).join('#'),
      transform: function transform(content) {
        return (0, _blocks.createBlock)(name, {
          level: level,
          content: content
        });
      }
    };
  })), (0, _toConsumableArray2.default)([1, 2, 3, 4, 5, 6].map(function (level) {
    return {
      type: 'enter',
      regExp: new RegExp("^/(h|H)".concat(level, "$")),
      transform: function transform(content) {
        return (0, _blocks.createBlock)(name, {
          level: level,
          content: content
        });
      }
    };
  }))),
  to: [{
    type: 'block',
    isMultiBlock: true,
    blocks: ['core/paragraph'],
    transform: function transform(attributes) {
      return attributes.map(function (_ref4) {
        var content = _ref4.content,
            anchor = _ref4.anchor;
        return (0, _blocks.createBlock)('core/paragraph', {
          content: content,
          anchor: anchor
        });
      });
    }
  }]
};
var _default = transforms;
exports.default = _default;
