// @wordpress/block-library/build/query-loop/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _edit = _interopRequireDefault(require("./edit"));

var _save = _interopRequireDefault(require("./save"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/query-loop",
  category: "design",
  usesContext: ["queryId", "query", "queryContext", "layout", "templateSlug"],
  supports: {
    reusable: false,
    html: false
  },
  style: "wp-block-query-loop",
  editorStyle: "wp-block-query-loop-editor"
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Query Loop', 'block title'),
  icon: _icons.loop,
  edit: _edit.default,
  save: _save.default,
  parent: ['core/query']
};
exports.settings = settings;
