// @wordpress/block-library/build/template-part/edit/selection/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TemplatePartSelection;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _blockEditor = require("@wordpress/block-editor");

var _keycodes = require("@wordpress/keycodes");

var _templatePartPreviews = _interopRequireDefault(require("./template-part-previews"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var preventArrowKeysPropagation = function preventArrowKeysPropagation(event) {
  if ([_keycodes.LEFT, _keycodes.DOWN, _keycodes.RIGHT, _keycodes.UP, _keycodes.BACKSPACE, _keycodes.ENTER].includes(event.keyCode)) {
    // Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
    event.stopPropagation();
  }
};

var stopKeyPropagation = function stopKeyPropagation(event) {
  return event.stopPropagation();
}; // Disable reason (no-static-element-interactions): Navigational key-presses within
// the menu are prevented from triggering WritingFlow and ObserveTyping interactions.

/* eslint-disable jsx-a11y/no-static-element-interactions */


function TemplatePartSelection(_ref) {
  var setAttributes = _ref.setAttributes,
      onClose = _ref.onClose;

  var _useState = (0, _element.useState)(''),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      filterValue = _useState2[0],
      setFilterValue = _useState2[1];

  return (0, _element.createElement)("div", {
    onKeyPress: stopKeyPropagation,
    onKeyDown: preventArrowKeysPropagation
  }, (0, _element.createElement)(_blockEditor.__experimentalSearchForm, {
    value: filterValue,
    onChange: setFilterValue,
    className: "wp-block-template-part__selection-preview-search-form"
  }), (0, _element.createElement)("div", {
    className: "wp-block-template-part__selection-preview-container"
  }, (0, _element.createElement)(_templatePartPreviews.default, {
    setAttributes: setAttributes,
    filterValue: filterValue,
    onClose: onClose
  })));
}
/* eslint-enable jsx-a11y/no-static-element-interactions */
