// @wordpress/block-library/build/template-part/edit/selection/template-part-previews.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TemplatePartPreviews;

var _element = require("@wordpress/element");

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _lodash = require("lodash");

var _data = require("@wordpress/data");

var _blocks = require("@wordpress/blocks");

var _keycodes = require("@wordpress/keycodes");

var _i18n = require("@wordpress/i18n");

var _blockEditor = require("@wordpress/block-editor");

var _components = require("@wordpress/components");

var _compose = require("@wordpress/compose");

var _notices = require("@wordpress/notices");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
function PreviewPlaceholder() {
  return (0, _element.createElement)("div", {
    className: "wp-block-template-part__selection-preview-item is-placeholder",
    tabIndex: 0
  });
}

function TemplatePartItem(_ref) {
  var templatePart = _ref.templatePart,
      setAttributes = _ref.setAttributes,
      onClose = _ref.onClose,
      composite = _ref.composite;
  var slug = templatePart.slug,
      theme = templatePart.theme,
      title = templatePart.title.rendered; // The 'raw' property is not defined for a brief period in the save cycle.
  // The fallback prevents an error in the parse function while saving.

  var content = templatePart.content.raw || '';
  var blocks = (0, _element.useMemo)(function () {
    return (0, _blocks.parse)(content);
  }, [content]);

  var _useDispatch = (0, _data.useDispatch)(_notices.store),
      createSuccessNotice = _useDispatch.createSuccessNotice;

  var onClick = (0, _element.useCallback)(function () {
    setAttributes({
      slug: slug,
      theme: theme
    });
    createSuccessNotice((0, _i18n.sprintf)(
    /* translators: %s: template part title. */
    (0, _i18n.__)('Template Part "%s" inserted.'), title || slug), {
      type: 'snackbar'
    });
    onClose();
  }, [slug, theme]);
  return (0, _element.createElement)(_components.__unstableCompositeItem, (0, _extends2.default)({
    as: "div",
    className: "wp-block-template-part__selection-preview-item",
    role: "option",
    onClick: onClick,
    onKeyDown: function onKeyDown(event) {
      if (_keycodes.ENTER === event.keyCode || _keycodes.SPACE === event.keyCode) {
        onClick();
      }
    },
    tabIndex: 0,
    "aria-label": title || slug
  }, composite), (0, _element.createElement)(_blockEditor.BlockPreview, {
    blocks: blocks
  }), (0, _element.createElement)("div", {
    className: "wp-block-template-part__selection-preview-item-title"
  }, title || slug));
}

function PanelGroup(_ref2) {
  var title = _ref2.title,
      icon = _ref2.icon,
      children = _ref2.children;
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)("div", {
    className: "wp-block-template-part__selection-panel-group-header"
  }, (0, _element.createElement)("span", {
    className: "wp-block-template-part__selection-panel-group-title"
  }, title), (0, _element.createElement)(_components.Icon, {
    icon: icon
  })), (0, _element.createElement)("div", {
    className: "wp-block-template-part__selection-panel-group-content"
  }, children));
}

function TemplatePartsByTheme(_ref3) {
  var templateParts = _ref3.templateParts,
      setAttributes = _ref3.setAttributes,
      onClose = _ref3.onClose,
      composite = _ref3.composite;
  var templatePartsByTheme = (0, _element.useMemo)(function () {
    return Object.values((0, _lodash.groupBy)(templateParts, 'theme'));
  }, [templateParts]);
  var currentShownTPs = (0, _compose.useAsyncList)(templateParts);
  return templatePartsByTheme.map(function (templatePartList) {
    return (0, _element.createElement)(PanelGroup, {
      key: templatePartList[0].theme
    }, templatePartList.map(function (templatePart) {
      return currentShownTPs.includes(templatePart) ? (0, _element.createElement)(TemplatePartItem, {
        key: templatePart.id,
        templatePart: templatePart,
        setAttributes: setAttributes,
        onClose: onClose,
        composite: composite
      }) : (0, _element.createElement)(PreviewPlaceholder, {
        key: templatePart.id
      });
    }));
  });
}

function TemplatePartSearchResults(_ref4) {
  var templateParts = _ref4.templateParts,
      setAttributes = _ref4.setAttributes,
      filterValue = _ref4.filterValue,
      onClose = _ref4.onClose,
      composite = _ref4.composite;
  var filteredTPs = (0, _element.useMemo)(function () {
    // Filter based on value.
    // Remove diacritics and convert to lowercase to normalize.
    var normalizedFilterValue = (0, _lodash.deburr)(filterValue).toLowerCase();
    var searchResults = templateParts.filter(function (_ref5) {
      var slug = _ref5.slug,
          theme = _ref5.theme;
      return slug.toLowerCase().includes(normalizedFilterValue) || // Since diacritics can be used in theme names, remove them for the comparison.
      (0, _lodash.deburr)(theme).toLowerCase().includes(normalizedFilterValue);
    }); // Order based on value location.

    searchResults.sort(function (a, b) {
      // First prioritize index found in slug.
      var indexInSlugA = a.slug.toLowerCase().indexOf(normalizedFilterValue);
      var indexInSlugB = b.slug.toLowerCase().indexOf(normalizedFilterValue);

      if (indexInSlugA !== -1 && indexInSlugB !== -1) {
        return indexInSlugA - indexInSlugB;
      } else if (indexInSlugA !== -1) {
        return -1;
      } else if (indexInSlugB !== -1) {
        return 1;
      } // Second prioritize index found in theme.
      // Since diacritics can be used in theme names, remove them for the comparison.


      return (0, _lodash.deburr)(a.theme).toLowerCase().indexOf(normalizedFilterValue) - (0, _lodash.deburr)(b.theme).toLowerCase().indexOf(normalizedFilterValue);
    });
    return searchResults;
  }, [filterValue, templateParts]);
  var currentShownTPs = (0, _compose.useAsyncList)(filteredTPs);
  return filteredTPs.map(function (templatePart) {
    return (0, _element.createElement)(PanelGroup, {
      key: templatePart.id,
      title: templatePart.theme || (0, _i18n.__)('Custom')
    }, currentShownTPs.includes(templatePart) ? (0, _element.createElement)(TemplatePartItem, {
      key: templatePart.id,
      templatePart: templatePart,
      setAttributes: setAttributes,
      onClose: onClose,
      composite: composite
    }) : (0, _element.createElement)(PreviewPlaceholder, {
      key: templatePart.id
    }));
  });
}

function TemplatePartPreviews(_ref6) {
  var setAttributes = _ref6.setAttributes,
      filterValue = _ref6.filterValue,
      onClose = _ref6.onClose;
  var composite = (0, _components.__unstableUseCompositeState)();
  var templateParts = (0, _data.useSelect)(function (select) {
    return select('core').getEntityRecords('postType', 'wp_template_part') || [];
  }, []);

  if (!templateParts || !templateParts.length) {
    return null;
  }

  if (filterValue) {
    return (0, _element.createElement)(_components.__unstableComposite, (0, _extends2.default)({}, composite, {
      role: "listbox",
      "aria-label": (0, _i18n.__)('List of template parts')
    }), (0, _element.createElement)(TemplatePartSearchResults, {
      templateParts: templateParts,
      setAttributes: setAttributes,
      filterValue: filterValue,
      onClose: onClose,
      composite: composite
    }));
  }

  return (0, _element.createElement)(_components.__unstableComposite, (0, _extends2.default)({}, composite, {
    role: "listbox",
    "aria-label": (0, _i18n.__)('List of template parts')
  }), (0, _element.createElement)(TemplatePartsByTheme, {
    templateParts: templateParts,
    setAttributes: setAttributes,
    onClose: onClose,
    composite: composite
  }));
}
