// @wordpress/block-library/build-module/template-part/edit/selection/index.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { createElement } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __experimentalSearchForm as SearchForm } from '@wordpress/block-editor';
import { useState } from '@wordpress/element';
import { LEFT, RIGHT, UP, DOWN, BACKSPACE, ENTER } from '@wordpress/keycodes';
/**
 * Internal dependencies
 */

import TemplatePartPreviews from './template-part-previews';

var preventArrowKeysPropagation = function preventArrowKeysPropagation(event) {
  if ([LEFT, DOWN, RIGHT, UP, BACKSPACE, ENTER].includes(event.keyCode)) {
    // Stop the key event from propagating up to ObserveTyping.startTypingInTextField.
    event.stopPropagation();
  }
};

var stopKeyPropagation = function stopKeyPropagation(event) {
  return event.stopPropagation();
}; // Disable reason (no-static-element-interactions): Navigational key-presses within
// the menu are prevented from triggering WritingFlow and ObserveTyping interactions.

/* eslint-disable jsx-a11y/no-static-element-interactions */


export default function TemplatePartSelection(_ref) {
  var setAttributes = _ref.setAttributes,
      onClose = _ref.onClose;

  var _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      filterValue = _useState2[0],
      setFilterValue = _useState2[1];

  return createElement("div", {
    onKeyPress: stopKeyPropagation,
    onKeyDown: preventArrowKeysPropagation
  }, createElement(SearchForm, {
    value: filterValue,
    onChange: setFilterValue,
    className: "wp-block-template-part__selection-preview-search-form"
  }), createElement("div", {
    className: "wp-block-template-part__selection-preview-container"
  }, createElement(TemplatePartPreviews, {
    setAttributes: setAttributes,
    filterValue: filterValue,
    onClose: onClose
  })));
}
/* eslint-enable jsx-a11y/no-static-element-interactions */
