// @wordpress/block-library/build-module/template-part/edit/selection/template-part-previews.js
import _extends from "@babel/runtime/helpers/esm/extends";
import { createElement, Fragment } from "@wordpress/element";

/**
 * External dependencies
 */
import { groupBy, deburr } from 'lodash';
/**
 * WordPress dependencies
 */

import { useSelect, useDispatch } from '@wordpress/data';
import { parse } from '@wordpress/blocks';
import { useMemo, useCallback } from '@wordpress/element';
import { ENTER, SPACE } from '@wordpress/keycodes';
import { __, sprintf } from '@wordpress/i18n';
import { BlockPreview } from '@wordpress/block-editor';
import { __unstableComposite as Composite, __unstableCompositeItem as CompositeItem, Icon, __unstableUseCompositeState as useCompositeState } from '@wordpress/components';
import { useAsyncList } from '@wordpress/compose';
import { store as noticesStore } from '@wordpress/notices';

function PreviewPlaceholder() {
  return createElement("div", {
    className: "wp-block-template-part__selection-preview-item is-placeholder",
    tabIndex: 0
  });
}

function TemplatePartItem(_ref) {
  var templatePart = _ref.templatePart,
      setAttributes = _ref.setAttributes,
      onClose = _ref.onClose,
      composite = _ref.composite;
  var slug = templatePart.slug,
      theme = templatePart.theme,
      title = templatePart.title.rendered; // The 'raw' property is not defined for a brief period in the save cycle.
  // The fallback prevents an error in the parse function while saving.

  var content = templatePart.content.raw || '';
  var blocks = useMemo(function () {
    return parse(content);
  }, [content]);

  var _useDispatch = useDispatch(noticesStore),
      createSuccessNotice = _useDispatch.createSuccessNotice;

  var onClick = useCallback(function () {
    setAttributes({
      slug: slug,
      theme: theme
    });
    createSuccessNotice(sprintf(
    /* translators: %s: template part title. */
    __('Template Part "%s" inserted.'), title || slug), {
      type: 'snackbar'
    });
    onClose();
  }, [slug, theme]);
  return createElement(CompositeItem, _extends({
    as: "div",
    className: "wp-block-template-part__selection-preview-item",
    role: "option",
    onClick: onClick,
    onKeyDown: function onKeyDown(event) {
      if (ENTER === event.keyCode || SPACE === event.keyCode) {
        onClick();
      }
    },
    tabIndex: 0,
    "aria-label": title || slug
  }, composite), createElement(BlockPreview, {
    blocks: blocks
  }), createElement("div", {
    className: "wp-block-template-part__selection-preview-item-title"
  }, title || slug));
}

function PanelGroup(_ref2) {
  var title = _ref2.title,
      icon = _ref2.icon,
      children = _ref2.children;
  return createElement(Fragment, null, createElement("div", {
    className: "wp-block-template-part__selection-panel-group-header"
  }, createElement("span", {
    className: "wp-block-template-part__selection-panel-group-title"
  }, title), createElement(Icon, {
    icon: icon
  })), createElement("div", {
    className: "wp-block-template-part__selection-panel-group-content"
  }, children));
}

function TemplatePartsByTheme(_ref3) {
  var templateParts = _ref3.templateParts,
      setAttributes = _ref3.setAttributes,
      onClose = _ref3.onClose,
      composite = _ref3.composite;
  var templatePartsByTheme = useMemo(function () {
    return Object.values(groupBy(templateParts, 'theme'));
  }, [templateParts]);
  var currentShownTPs = useAsyncList(templateParts);
  return templatePartsByTheme.map(function (templatePartList) {
    return createElement(PanelGroup, {
      key: templatePartList[0].theme
    }, templatePartList.map(function (templatePart) {
      return currentShownTPs.includes(templatePart) ? createElement(TemplatePartItem, {
        key: templatePart.id,
        templatePart: templatePart,
        setAttributes: setAttributes,
        onClose: onClose,
        composite: composite
      }) : createElement(PreviewPlaceholder, {
        key: templatePart.id
      });
    }));
  });
}

function TemplatePartSearchResults(_ref4) {
  var templateParts = _ref4.templateParts,
      setAttributes = _ref4.setAttributes,
      filterValue = _ref4.filterValue,
      onClose = _ref4.onClose,
      composite = _ref4.composite;
  var filteredTPs = useMemo(function () {
    // Filter based on value.
    // Remove diacritics and convert to lowercase to normalize.
    var normalizedFilterValue = deburr(filterValue).toLowerCase();
    var searchResults = templateParts.filter(function (_ref5) {
      var slug = _ref5.slug,
          theme = _ref5.theme;
      return slug.toLowerCase().includes(normalizedFilterValue) || // Since diacritics can be used in theme names, remove them for the comparison.
      deburr(theme).toLowerCase().includes(normalizedFilterValue);
    }); // Order based on value location.

    searchResults.sort(function (a, b) {
      // First prioritize index found in slug.
      var indexInSlugA = a.slug.toLowerCase().indexOf(normalizedFilterValue);
      var indexInSlugB = b.slug.toLowerCase().indexOf(normalizedFilterValue);

      if (indexInSlugA !== -1 && indexInSlugB !== -1) {
        return indexInSlugA - indexInSlugB;
      } else if (indexInSlugA !== -1) {
        return -1;
      } else if (indexInSlugB !== -1) {
        return 1;
      } // Second prioritize index found in theme.
      // Since diacritics can be used in theme names, remove them for the comparison.


      return deburr(a.theme).toLowerCase().indexOf(normalizedFilterValue) - deburr(b.theme).toLowerCase().indexOf(normalizedFilterValue);
    });
    return searchResults;
  }, [filterValue, templateParts]);
  var currentShownTPs = useAsyncList(filteredTPs);
  return filteredTPs.map(function (templatePart) {
    return createElement(PanelGroup, {
      key: templatePart.id,
      title: templatePart.theme || __('Custom')
    }, currentShownTPs.includes(templatePart) ? createElement(TemplatePartItem, {
      key: templatePart.id,
      templatePart: templatePart,
      setAttributes: setAttributes,
      onClose: onClose,
      composite: composite
    }) : createElement(PreviewPlaceholder, {
      key: templatePart.id
    }));
  });
}

export default function TemplatePartPreviews(_ref6) {
  var setAttributes = _ref6.setAttributes,
      filterValue = _ref6.filterValue,
      onClose = _ref6.onClose;
  var composite = useCompositeState();
  var templateParts = useSelect(function (select) {
    return select('core').getEntityRecords('postType', 'wp_template_part') || [];
  }, []);

  if (!templateParts || !templateParts.length) {
    return null;
  }

  if (filterValue) {
    return createElement(Composite, _extends({}, composite, {
      role: "listbox",
      "aria-label": __('List of template parts')
    }), createElement(TemplatePartSearchResults, {
      templateParts: templateParts,
      setAttributes: setAttributes,
      filterValue: filterValue,
      onClose: onClose,
      composite: composite
    }));
  }

  return createElement(Composite, _extends({}, composite, {
    role: "listbox",
    "aria-label": __('List of template parts')
  }), createElement(TemplatePartsByTheme, {
    templateParts: templateParts,
    setAttributes: setAttributes,
    onClose: onClose,
    composite: composite
  }));
}
