// @wordpress/block-library/build/query-pagination-next/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _edit = _interopRequireDefault(require("./edit"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/query-pagination-next",
  category: "design",
  attributes: {
    label: {
      type: "string"
    }
  },
  usesContext: ["queryId", "query"],
  supports: {
    reusable: false,
    html: false,
    color: {
      gradients: true,
      link: true
    },
    fontSize: true,
    lineHeight: true
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Query Pagination Next', 'block title'),
  description: (0, _i18n.__)('Displays the next posts page link.'),
  icon: _icons.queryPaginationNext,
  edit: _edit.default,
  parent: ['core/query-pagination']
};
exports.settings = settings;
