// @wordpress/block-library/build/categories/edit.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CategoriesEdit;

var _element = require("@wordpress/element");

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _lodash = require("lodash");

var _components = require("@wordpress/components");

var _compose = require("@wordpress/compose");

var _data = require("@wordpress/data");

var _blockEditor = require("@wordpress/block-editor");

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */
function CategoriesEdit(_ref) {
  var _ref$attributes = _ref.attributes,
      displayAsDropdown = _ref$attributes.displayAsDropdown,
      showHierarchy = _ref$attributes.showHierarchy,
      showPostCounts = _ref$attributes.showPostCounts,
      setAttributes = _ref.setAttributes;
  var selectId = (0, _compose.useInstanceId)(CategoriesEdit, 'blocks-category-select');

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select('core'),
        getEntityRecords = _select.getEntityRecords;

    var _select2 = select('core/data'),
        isResolving = _select2.isResolving;

    var query = {
      per_page: -1,
      hide_empty: true
    };
    return {
      categories: getEntityRecords('taxonomy', 'category', query),
      isRequesting: isResolving('core', 'getEntityRecords', ['taxonomy', 'category', query])
    };
  }, []),
      categories = _useSelect.categories,
      isRequesting = _useSelect.isRequesting;

  var getCategoriesList = function getCategoriesList(parentId) {
    if (!(categories !== null && categories !== void 0 && categories.length)) {
      return [];
    }

    if (parentId === null) {
      return categories;
    }

    return categories.filter(function (_ref2) {
      var parent = _ref2.parent;
      return parent === parentId;
    });
  };

  var getCategoryListClassName = function getCategoryListClassName(level) {
    return "wp-block-categories__list wp-block-categories__list-level-".concat(level);
  };

  var toggleAttribute = function toggleAttribute(attributeName) {
    return function (newValue) {
      return setAttributes((0, _defineProperty2.default)({}, attributeName, newValue));
    };
  };

  var renderCategoryName = function renderCategoryName(name) {
    return !name ? (0, _i18n.__)('(Untitled)') : (0, _lodash.unescape)(name).trim();
  };

  var renderCategoryList = function renderCategoryList() {
    var parentId = showHierarchy ? 0 : null;
    var categoriesList = getCategoriesList(parentId);
    return (0, _element.createElement)("ul", {
      className: getCategoryListClassName(0)
    }, categoriesList.map(function (category) {
      return renderCategoryListItem(category, 0);
    }));
  };

  var renderCategoryListItem = function renderCategoryListItem(category, level) {
    var childCategories = getCategoriesList(category.id);
    var id = category.id,
        link = category.link,
        count = category.count,
        name = category.name;
    return (0, _element.createElement)("li", {
      key: id
    }, (0, _element.createElement)("a", {
      href: link,
      target: "_blank",
      rel: "noreferrer noopener"
    }, renderCategoryName(name)), showPostCounts && (0, _element.createElement)("span", {
      className: "wp-block-categories__post-count"
    }, " (".concat(count, ")")), showHierarchy && !!childCategories.length && (0, _element.createElement)("ul", {
      className: getCategoryListClassName(level + 1)
    }, childCategories.map(function (childCategory) {
      return renderCategoryListItem(childCategory, level + 1);
    })));
  };

  var renderCategoryDropdown = function renderCategoryDropdown() {
    var parentId = showHierarchy ? 0 : null;
    var categoriesList = getCategoriesList(parentId);
    return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_components.VisuallyHidden, {
      as: "label",
      htmlFor: selectId
    }, (0, _i18n.__)('Categories')), (0, _element.createElement)("select", {
      id: selectId,
      className: "wp-block-categories__dropdown"
    }, categoriesList.map(function (category) {
      return renderCategoryDropdownItem(category, 0);
    })));
  };

  var renderCategoryDropdownItem = function renderCategoryDropdownItem(category, level) {
    var id = category.id,
        count = category.count,
        name = category.name;
    var childCategories = getCategoriesList(id);
    return [(0, _element.createElement)("option", {
      key: id
    }, (0, _lodash.times)(level * 3, function () {
      return '\xa0';
    }), renderCategoryName(name), showPostCounts && " (".concat(count, ")")), showHierarchy && !!childCategories.length && childCategories.map(function (childCategory) {
      return renderCategoryDropdownItem(childCategory, level + 1);
    })];
  };

  return (0, _element.createElement)("div", (0, _blockEditor.useBlockProps)(), (0, _element.createElement)(_blockEditor.InspectorControls, null, (0, _element.createElement)(_components.PanelBody, {
    title: (0, _i18n.__)('Categories settings')
  }, (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Display as dropdown'),
    checked: displayAsDropdown,
    onChange: toggleAttribute('displayAsDropdown')
  }), (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Show hierarchy'),
    checked: showHierarchy,
    onChange: toggleAttribute('showHierarchy')
  }), (0, _element.createElement)(_components.ToggleControl, {
    label: (0, _i18n.__)('Show post counts'),
    checked: showPostCounts,
    onChange: toggleAttribute('showPostCounts')
  }))), isRequesting && (0, _element.createElement)(_components.Placeholder, {
    icon: _icons.pin,
    label: (0, _i18n.__)('Categories')
  }, (0, _element.createElement)(_components.Spinner, null)), !isRequesting && (displayAsDropdown ? renderCategoryDropdown() : renderCategoryList()));
}
