// @wordpress/editor/build/components/post-format/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = PostFormat;
exports.POST_FORMATS = void 0;

var _element = require("@wordpress/element");

var _lodash = require("lodash");

var _i18n = require("@wordpress/i18n");

var _components = require("@wordpress/components");

var _data = require("@wordpress/data");

var _compose = require("@wordpress/compose");

var _check = _interopRequireDefault(require("./check"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
// All WP post formats, sorted alphabetically by translated name.
var POST_FORMATS = [{
  id: 'aside',
  caption: (0, _i18n.__)('Aside')
}, {
  id: 'audio',
  caption: (0, _i18n.__)('Audio')
}, {
  id: 'chat',
  caption: (0, _i18n.__)('Chat')
}, {
  id: 'gallery',
  caption: (0, _i18n.__)('Gallery')
}, {
  id: 'image',
  caption: (0, _i18n.__)('Image')
}, {
  id: 'link',
  caption: (0, _i18n.__)('Link')
}, {
  id: 'quote',
  caption: (0, _i18n.__)('Quote')
}, {
  id: 'standard',
  caption: (0, _i18n.__)('Standard')
}, {
  id: 'status',
  caption: (0, _i18n.__)('Status')
}, {
  id: 'video',
  caption: (0, _i18n.__)('Video')
}].sort(function (a, b) {
  var normalizedA = a.caption.toUpperCase();
  var normalizedB = b.caption.toUpperCase();

  if (normalizedA < normalizedB) {
    return -1;
  }

  if (normalizedA > normalizedB) {
    return 1;
  }

  return 0;
});
exports.POST_FORMATS = POST_FORMATS;

function PostFormat() {
  var instanceId = (0, _compose.useInstanceId)(PostFormat);
  var postFormatSelectorId = "post-format-selector-".concat(instanceId);

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select('core/editor'),
        getEditedPostAttribute = _select.getEditedPostAttribute,
        getSuggestedPostFormat = _select.getSuggestedPostFormat;

    var _postFormat = getEditedPostAttribute('format');

    var themeSupports = select('core').getThemeSupports();
    return {
      postFormat: _postFormat !== null && _postFormat !== void 0 ? _postFormat : 'standard',
      suggestedFormat: getSuggestedPostFormat(),
      // Ensure current format is always in the set.
      // The current format may not be a format supported by the theme.
      supportedFormats: (0, _lodash.union)([_postFormat], (0, _lodash.get)(themeSupports, ['formats'], []))
    };
  }, []),
      postFormat = _useSelect.postFormat,
      suggestedFormat = _useSelect.suggestedFormat,
      supportedFormats = _useSelect.supportedFormats;

  var formats = POST_FORMATS.filter(function (format) {
    return (0, _lodash.includes)(supportedFormats, format.id);
  });
  var suggestion = (0, _lodash.find)(formats, function (format) {
    return format.id === suggestedFormat;
  });

  var _useDispatch = (0, _data.useDispatch)('core/editor'),
      editPost = _useDispatch.editPost;

  var onUpdatePostFormat = function onUpdatePostFormat(format) {
    return editPost({
      format: format
    });
  };

  return (0, _element.createElement)(_check.default, null, (0, _element.createElement)("div", {
    className: "editor-post-format"
  }, (0, _element.createElement)("div", {
    className: "editor-post-format__content"
  }, (0, _element.createElement)("label", {
    htmlFor: postFormatSelectorId
  }, (0, _i18n.__)('Post Format')), (0, _element.createElement)(_components.SelectControl, {
    value: postFormat,
    onChange: function onChange(format) {
      return onUpdatePostFormat(format);
    },
    id: postFormatSelectorId,
    options: formats.map(function (format) {
      return {
        label: format.caption,
        value: format.id
      };
    })
  })), suggestion && suggestion.id !== postFormat && (0, _element.createElement)("div", {
    className: "editor-post-format__suggestion"
  }, (0, _i18n.__)('Suggestion:'), ' ', (0, _element.createElement)(_components.Button, {
    isLink: true,
    onClick: function onClick() {
      return onUpdatePostFormat(suggestion.id);
    }
  }, suggestion.caption))));
}
