// @wordpress/block-library/build/post-author/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _edit = _interopRequireDefault(require("./edit"));

var _icon = _interopRequireDefault(require("./icon"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/post-author",
  category: "design",
  attributes: {
    textAlign: {
      type: "string"
    },
    avatarSize: {
      type: "number",
      "default": 48
    },
    showAvatar: {
      type: "boolean",
      "default": true
    },
    showBio: {
      type: "boolean"
    },
    byline: {
      type: "string"
    }
  },
  usesContext: ["postType", "postId"],
  supports: {
    html: false,
    fontSize: true,
    color: {
      gradients: true,
      link: true
    },
    lineHeight: true
  },
  editorStyle: "wp-block-post-author-editor",
  style: "wp-block-post-author"
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Post Author', 'block title'),
  description: (0, _i18n.__)('Add the author of this post.'),
  icon: _icon.default,
  edit: _edit.default
};
exports.settings = settings;
