// @wordpress/block-library/build-module/template-part/edit/placeholder/index.js
import _regeneratorRuntime from "@babel/runtime/regenerator";
import _asyncToGenerator from "@babel/runtime/helpers/esm/asyncToGenerator";
import { createElement, Fragment } from "@wordpress/element";

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';
import { useCallback, useEffect } from '@wordpress/element';
import { useDispatch } from '@wordpress/data';
import { Placeholder, Dropdown, Button, Spinner } from '@wordpress/components';
import { blockDefault } from '@wordpress/icons';
import { serialize } from '@wordpress/blocks';
/**
 * Internal dependencies
 */

import TemplatePartSelection from '../selection';
export default function TemplatePartPlaceholder(_ref) {
  var setAttributes = _ref.setAttributes,
      innerBlocks = _ref.innerBlocks;

  var _useDispatch = useDispatch('core'),
      saveEntityRecord = _useDispatch.saveEntityRecord;

  var onCreate = useCallback( /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
    var title, templatePart;
    return _regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            title = __('Untitled Template Part');
            _context.next = 3;
            return saveEntityRecord('postType', 'wp_template_part', {
              title: title,
              slug: 'template-part',
              content: serialize(innerBlocks)
            });

          case 3:
            templatePart = _context.sent;
            setAttributes({
              slug: templatePart.slug,
              theme: templatePart.theme
            });

          case 5:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  })), [setAttributes]); // If there are inner blocks present, the content for creation is clear.
  // Therefore immediately create the template part with the given inner blocks as its content.

  useEffect(function () {
    if (innerBlocks.length) {
      onCreate();
    }
  }, []);

  if (innerBlocks.length) {
    return createElement(Placeholder, null, createElement(Spinner, null));
  }

  return createElement(Placeholder, {
    icon: blockDefault,
    label: __('Template Part'),
    instructions: __('Create a new template part or pick an existing one from the list.')
  }, createElement(Dropdown, {
    contentClassName: "wp-block-template-part__placeholder-preview-dropdown-content",
    position: "bottom right left",
    renderToggle: function renderToggle(_ref3) {
      var isOpen = _ref3.isOpen,
          onToggle = _ref3.onToggle;
      return createElement(Fragment, null, createElement(Button, {
        isPrimary: true,
        onClick: onToggle,
        "aria-expanded": isOpen
      }, __('Choose existing')), createElement(Button, {
        isTertiary: true,
        onClick: onCreate
      }, __('New template part')));
    },
    renderContent: function renderContent(_ref4) {
      var onClose = _ref4.onClose;
      return createElement(TemplatePartSelection, {
        setAttributes: setAttributes,
        onClose: onClose
      });
    }
  }));
}
