// @wordpress/block-library/build/buttons/edit.native.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = ButtonsEdit;

var _element = require("@wordpress/element");

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lodash = require("lodash");

var _reactNative = require("react-native");

var _blockEditor = require("@wordpress/block-editor");

var _blocks = require("@wordpress/blocks");

var _compose = require("@wordpress/compose");

var _data = require("@wordpress/data");

var _components = require("@wordpress/components");

var _button = require("../button/");

var _editor = _interopRequireDefault(require("./editor.scss"));

var _contentJustificationDropdown = _interopRequireDefault(require("./content-justification-dropdown"));

/**
 * External dependencies
 */

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var ALLOWED_BLOCKS = [_button.name];
var BUTTONS_TEMPLATE = [['core/button']];

function ButtonsEdit(_ref) {
  var contentJustification = _ref.attributes.contentJustification,
      clientId = _ref.clientId,
      isSelected = _ref.isSelected,
      setAttributes = _ref.setAttributes,
      blockWidth = _ref.blockWidth;

  var _useResizeObserver = (0, _compose.useResizeObserver)(),
      _useResizeObserver2 = (0, _slicedToArray2.default)(_useResizeObserver, 2),
      resizeObserver = _useResizeObserver2[0],
      sizes = _useResizeObserver2[1];

  var _useState = (0, _element.useState)(0),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      maxWidth = _useState2[0],
      setMaxWidth = _useState2[1];

  var spacing = _editor.default.spacing.marginLeft;

  var _useSelect = (0, _data.useSelect)(function (select) {
    var _select = select('core/block-editor'),
        getBlockCount = _select.getBlockCount,
        _getBlockOrder = _select.getBlockOrder,
        getBlockParents = _select.getBlockParents,
        getSelectedBlockClientId = _select.getSelectedBlockClientId;

    var selectedBlockClientId = getSelectedBlockClientId();
    var selectedBlockParents = getBlockParents(selectedBlockClientId, true);
    return {
      getBlockOrder: _getBlockOrder,
      isInnerButtonSelected: selectedBlockParents[0] === clientId,
      // The purpose of `shouldDelete` check is giving the ability to
      // pass to mobile toolbar function called `onDelete` which removes
      // the whole `Buttons` container along with the last inner button
      // when there is exactly one button.
      shouldDelete: getBlockCount(clientId) === 1
    };
  }, [clientId]),
      getBlockOrder = _useSelect.getBlockOrder,
      isInnerButtonSelected = _useSelect.isInnerButtonSelected,
      shouldDelete = _useSelect.shouldDelete;

  var _useDispatch = (0, _data.useDispatch)('core/block-editor'),
      insertBlock = _useDispatch.insertBlock,
      removeBlock = _useDispatch.removeBlock,
      selectBlock = _useDispatch.selectBlock;

  (0, _element.useEffect)(function () {
    var margins = 2 * _editor.default.parent.marginRight;

    var _ref2 = sizes || {},
        width = _ref2.width;

    if (width) {
      setMaxWidth(width - margins);
    }
  }, [sizes]);
  var onAddNextButton = (0, _lodash.debounce)(function (selectedId) {
    var order = getBlockOrder(clientId);
    var selectedButtonIndex = order.findIndex(function (i) {
      return i === selectedId;
    });
    var index = selectedButtonIndex === -1 ? order.length + 1 : selectedButtonIndex;
    var insertedBlock = (0, _blocks.createBlock)('core/button');
    insertBlock(insertedBlock, index, clientId);
    selectBlock(insertedBlock.clientId);
  }, 200);

  function onChangeContentJustification(updatedValue) {
    setAttributes({
      contentJustification: updatedValue
    });
  }

  var renderFooterAppender = (0, _element.useRef)(function () {
    return (0, _element.createElement)(_reactNative.View, {
      style: _editor.default.appenderContainer
    }, (0, _element.createElement)(_blockEditor.InnerBlocks.ButtonBlockAppender, {
      isFloating: true,
      onAddBlock: onAddNextButton
    }));
  });
  var shouldRenderFooterAppender = isSelected || isInnerButtonSelected;
  return (0, _element.createElement)(_element.Fragment, null, (0, _element.createElement)(_blockEditor.BlockControls, null, (0, _element.createElement)(_components.ToolbarGroup, null, (0, _element.createElement)(_components.ToolbarItem, null, function (toggleProps) {
    return (0, _element.createElement)(_contentJustificationDropdown.default, {
      toggleProps: toggleProps,
      value: contentJustification,
      onChange: onChangeContentJustification
    });
  }))), resizeObserver, (0, _element.createElement)(_blockEditor.InnerBlocks, {
    allowedBlocks: ALLOWED_BLOCKS,
    template: BUTTONS_TEMPLATE,
    renderFooterAppender: shouldRenderFooterAppender && renderFooterAppender.current,
    orientation: "horizontal",
    horizontalAlignment: contentJustification,
    onDeleteBlock: shouldDelete ? function () {
      return removeBlock(clientId);
    } : undefined,
    onAddBlock: onAddNextButton,
    parentWidth: maxWidth,
    marginHorizontal: spacing,
    marginVertical: spacing,
    __experimentalLayout: {
      type: 'default',
      alignments: []
    },
    templateInsertUpdatesSelection: true,
    blockWidth: blockWidth
  }));
}
