// @wordpress/block-library/build/buttons/transforms.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _blocks = require("@wordpress/blocks");

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var _apiVersion$name$cate = {
  apiVersion: 2,
  name: "core/buttons",
  category: "design",
  attributes: {
    contentJustification: {
      type: "string"
    },
    orientation: {
      type: "string",
      "default": "horizontal"
    }
  },
  supports: {
    anchor: true,
    align: ["wide", "full"]
  },
  editorStyle: "wp-block-buttons-editor",
  style: "wp-block-buttons"
},
    name = _apiVersion$name$cate.name;
var transforms = {
  from: [{
    type: 'block',
    isMultiBlock: true,
    blocks: ['core/button'],
    transform: function transform(buttons) {
      return (// Creates the buttons block
        (0, _blocks.createBlock)(name, {}, // Loop the selected buttons
        buttons.map(function (attributes) {
          return (// Create singular button in the buttons block
            (0, _blocks.createBlock)('core/button', attributes)
          );
        }))
      );
    }
  }]
};
var _default = transforms;
exports.default = _default;
