// @wordpress/block-library/build-module/query/index.js
/**
 * WordPress dependencies
 */
import { __, _x } from '@wordpress/i18n';
import { loop as icon } from '@wordpress/icons';
/**
 * Internal dependencies
 */

var metadata = {
  apiVersion: 2,
  name: "core/query",
  category: "design",
  attributes: {
    queryId: {
      type: "number"
    },
    query: {
      type: "object",
      "default": {
        perPage: null,
        pages: 0,
        offset: 0,
        postType: "post",
        categoryIds: [],
        tagIds: [],
        order: "desc",
        orderBy: "date",
        author: "",
        search: "",
        exclude: [],
        sticky: "",
        inherit: true
      }
    },
    layout: {
      type: "object",
      "default": {
        type: "list"
      }
    }
  },
  providesContext: {
    queryId: "queryId",
    query: "query",
    layout: "layout"
  },
  usesContext: ["postId"],
  supports: {
    html: false
  },
  editorStyle: "wp-block-query-editor"
};
import edit from './edit';
import save from './save';
import variations from './variations';
var name = metadata.name;
export { metadata, name };
export var settings = {
  title: _x('Query', 'block title'),
  icon: icon,
  description: __('Displays a list of posts as a result of a query.'),
  edit: edit,
  save: save,
  variations: variations
};
export { useQueryContext } from './edit';
