// @wordpress/block-library/build-module/buttons/content-justification-dropdown.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { createElement } from "@wordpress/element";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

/**
 * WordPress dependencies
 */
import { DropdownMenu } from '@wordpress/components';
import { __ } from '@wordpress/i18n';
import { justifyLeft, justifyCenter, justifyRight } from '@wordpress/icons';
var DEFAULT_ALLOWED_VALUES = ['left', 'center', 'right'];
var CONTROLS = {
  left: {
    icon: justifyLeft,
    title: __('Justify content left')
  },
  center: {
    icon: justifyCenter,
    title: __('Justify content center')
  },
  right: {
    icon: justifyRight,
    title: __('Justify content right')
  }
};
var DEFAULT_ICON = CONTROLS.left.icon;
/**
 * Dropdown for selecting a content justification option.
 *
 * @param {Object}   props                 Component props.
 * @param {string[]} [props.allowedValues] List of options to include. Default:
 *                                         ['left', 'center', 'right'].
 * @param {()=>void} props.onChange        Callback to run when an option is
 *                                         selected in the dropdown.
 * @param {Object}   props.toggleProps     Props to pass to the dropdown toggle.
 * @param {string}   props.value           The current content justification
 *                                         value.
 *
 * @return {WPComponent} The component.
 */

export default function ContentJustificationDropdown(_ref) {
  var _CONTROLS$value$icon, _CONTROLS$value;

  var onChange = _ref.onChange,
      _ref$allowedValues = _ref.allowedValues,
      allowedValues = _ref$allowedValues === void 0 ? DEFAULT_ALLOWED_VALUES : _ref$allowedValues,
      toggleProps = _ref.toggleProps,
      value = _ref.value;
  return createElement(DropdownMenu, {
    icon: (_CONTROLS$value$icon = (_CONTROLS$value = CONTROLS[value]) === null || _CONTROLS$value === void 0 ? void 0 : _CONTROLS$value.icon) !== null && _CONTROLS$value$icon !== void 0 ? _CONTROLS$value$icon : DEFAULT_ICON,
    label: __('Change content justification'),
    controls: allowedValues.map(function (allowedValue) {
      return _objectSpread(_objectSpread({}, CONTROLS[allowedValue]), {}, {
        isActive: value === allowedValue,
        role: 'menuitemradio',
        onClick: function onClick() {
          return onChange(value === allowedValue ? undefined : allowedValue);
        }
      });
    }),
    toggleProps: toggleProps
  });
}
