// @wordpress/block-library/build-module/buttons/index.js
/**
 * WordPress dependencies
 */
import { __, _x } from '@wordpress/i18n';
import { button as icon } from '@wordpress/icons';
/**
 * Internal dependencies
 */

import deprecated from './deprecated';
import transforms from './transforms';
import edit from './edit';
var metadata = {
  apiVersion: 2,
  name: "core/buttons",
  category: "design",
  attributes: {
    contentJustification: {
      type: "string"
    },
    orientation: {
      type: "string",
      "default": "horizontal"
    }
  },
  supports: {
    anchor: true,
    align: ["wide", "full"]
  },
  editorStyle: "wp-block-buttons-editor",
  style: "wp-block-buttons"
};
import save from './save';
import variations from './variations';
var name = metadata.name;
export { metadata, name };
export var settings = {
  title: _x('Buttons', 'block title'),
  description: __('Prompt visitors to take action with a group of button-style links.'),
  icon: icon,
  keywords: [__('link')],
  example: {
    innerBlocks: [{
      name: 'core/button',
      attributes: {
        text: __('Find out more')
      }
    }, {
      name: 'core/button',
      attributes: {
        text: __('Contact us')
      }
    }]
  },
  deprecated: deprecated,
  transforms: transforms,
  edit: edit,
  save: save,
  variations: variations
};
