// @wordpress/editor/build/components/post-author/select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _element = require("@wordpress/element");

var _i18n = require("@wordpress/i18n");

var _data = require("@wordpress/data");

var _htmlEntities = require("@wordpress/html-entities");

var _components = require("@wordpress/components");

/**
 * WordPress dependencies
 */
function PostAuthorSelect() {
  var _useDispatch = (0, _data.useDispatch)('core/editor'),
      editPost = _useDispatch.editPost;

  var _useSelect = (0, _data.useSelect)(function (select) {
    var authorsFromAPI = select('core').getAuthors();
    return {
      postAuthor: select('core/editor').getEditedPostAttribute('author'),
      authors: authorsFromAPI.map(function (author) {
        return {
          label: (0, _htmlEntities.decodeEntities)(author.name),
          value: author.id
        };
      })
    };
  }, []),
      postAuthor = _useSelect.postAuthor,
      authors = _useSelect.authors;

  var setAuthorId = function setAuthorId(value) {
    var author = Number(value);
    editPost({
      author: author
    });
  };

  return (0, _element.createElement)(_components.SelectControl, {
    className: "post-author-selector",
    label: (0, _i18n.__)('Author'),
    options: authors,
    onChange: setAuthorId,
    value: postAuthor
  });
}

var _default = PostAuthorSelect;
exports.default = _default;
