// @wordpress/block-library/build/post-title/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.settings = exports.name = exports.metadata = void 0;

var _i18n = require("@wordpress/i18n");

var _icons = require("@wordpress/icons");

var _edit = _interopRequireDefault(require("./edit"));

/**
 * WordPress dependencies
 */

/**
 * Internal dependencies
 */
var metadata = {
  apiVersion: 2,
  name: "core/post-title",
  category: "design",
  usesContext: ["postId", "postType"],
  attributes: {
    textAlign: {
      type: "string"
    },
    level: {
      type: "number",
      "default": 2
    },
    isLink: {
      type: "boolean",
      "default": false
    },
    rel: {
      type: "string",
      attribute: "rel",
      "default": ""
    },
    linkTarget: {
      type: "string",
      "default": "_self"
    }
  },
  supports: {
    align: ["wide", "full"],
    html: false,
    color: {
      gradients: true
    },
    fontSize: true,
    lineHeight: true,
    __experimentalFontFamily: true,
    __experimentalSelector: {
      "core/post-title/h1": {
        title: "h1",
        selector: "h1.wp-block-post-title",
        attributes: {
          level: 1
        }
      },
      "core/post-title/h2": {
        title: "h2",
        selector: "h2.wp-block-post-title",
        attributes: {
          level: 2
        }
      },
      "core/post-title/h3": {
        title: "h3",
        selector: "h3.wp-block-post-title",
        attributes: {
          level: 3
        }
      },
      "core/post-title/h4": {
        title: "h4",
        selector: "h4.wp-block-post-title",
        attributes: {
          level: 4
        }
      },
      "core/post-title/h5": {
        title: "h5",
        selector: "h5.wp-block-post-title",
        attributes: {
          level: 5
        }
      },
      "core/post-title/h6": {
        title: "h6",
        selector: "h6.wp-block-post-title",
        attributes: {
          level: 6
        }
      }
    }
  }
};
exports.metadata = metadata;
var name = metadata.name;
exports.name = name;
var settings = {
  title: (0, _i18n._x)('Post Title', 'block title'),
  description: (0, _i18n.__)('Displays the title of a post, page, or any other content-type.'),
  icon: _icons.postTitle,
  edit: _edit.default
};
exports.settings = settings;
