// @wordpress/block-library/build-module/rss/index.js
/**
 * WordPress dependencies
 */
import { rss as icon } from '@wordpress/icons';
import { __, _x } from '@wordpress/i18n';
/**
 * Internal dependencies
 */

var metadata = {
  apiVersion: 2,
  name: "core/rss",
  category: "widgets",
  attributes: {
    columns: {
      type: "number",
      "default": 2
    },
    blockLayout: {
      type: "string",
      "default": "list"
    },
    feedURL: {
      type: "string",
      "default": ""
    },
    itemsToShow: {
      type: "number",
      "default": 5
    },
    displayExcerpt: {
      type: "boolean",
      "default": false
    },
    displayAuthor: {
      type: "boolean",
      "default": false
    },
    displayDate: {
      type: "boolean",
      "default": false
    },
    excerptLength: {
      type: "number",
      "default": 55
    }
  },
  supports: {
    align: true,
    html: false
  },
  editorStyle: "wp-block-rss-editor",
  style: "wp-block-rss"
};
import edit from './edit';
var name = metadata.name;
export { metadata, name };
export var settings = {
  title: _x('RSS', 'block title'),
  description: __('Display entries from any RSS or Atom feed.'),
  icon: icon,
  keywords: [__('atom'), __('feed')],
  example: {
    attributes: {
      feedURL: 'https://wordpress.org'
    }
  },
  edit: edit
};
